{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import datetime\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEFAULT_NAME_COLUMN_DATE = 'Date'\n",
    "DEFAULT_NAME_COLUMN_DESCRIPTION = 'Description'\n",
    "DEFAULT_NAME_COLUMN_TOTAL = 'Totals'\n",
    "DEFAULT_NAME_COLUMN_COUNTRY = 'Country'\n",
    "\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES = 'New cases'\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS = 'New deaths'\n",
    "\n",
    "DEFAULT_DATA_FORMAT = '%Y-%m-%d'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Auxiliary functions"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def sanitize_date(date_col, country_index, original_formats):\n",
    "    date_list = list()\n",
    "    for i, date in date_col.iteritems():\n",
    "        date_rep = None\n",
    "        try:\n",
    "            date_rep = datetime.datetime.strptime(date, original_formats[country_index][0])\n",
    "        except ValueError:\n",
    "            date_rep = datetime.datetime.strptime(date, original_formats[country_index][1])\n",
    "        date_list.append(date_rep.strftime(DEFAULT_DATA_FORMAT))\n",
    "    return np.array(date_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sanitize_date(date, original_formats):\n",
    "    '''\n",
    "    Transform a date specified in a string with a certain original\n",
    "    format into a string with the date in the default format.\n",
    "    '''\n",
    "    date_rep = None\n",
    "    try:\n",
    "        date_rep = datetime.datetime.strptime(date, original_formats[0])\n",
    "    except ValueError:\n",
    "        date_rep = datetime.datetime.strptime(date, original_formats[1])\n",
    "    return date_rep.strftime(DEFAULT_DATA_FORMAT)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def sum_string_list(ls):\n",
    "    temp_sum = 0\n",
    "    for elem in ls:\n",
    "        temp_sum += float(elem)\n",
    "    return temp_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_daily_new_deaths(np_list):\n",
    "    '''\n",
    "    Given a list of accumulated data, calculate the data for each day.\n",
    "    It assumes ordered data.\n",
    "    '''\n",
    "    last_known_n = -1\n",
    "    final_list = list()\n",
    "    for i, elem in np_list.iteritems():\n",
    "        if (math.isnan(elem)):\n",
    "            final_list.append(0)\n",
    "        else:\n",
    "            if (last_known_n == -1):\n",
    "                final_list.append(0)\n",
    "            else:\n",
    "                final_list.append(elem - last_known_n)\n",
    "            last_known_n = elem\n",
    "    return np.array(final_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "country_keys = ['guinea', 'liberia', 'sl']\n",
    "paths = {x: DATA_FOLDER + '/ebola/'+ x + '_data' for x in country_keys}\n",
    "countries = {'guinea': 'Guinea', 'liberia': 'Liberia', 'sl': 'Sierra Leone'}\n",
    "\n",
    "wanted_columns = {'guinea': ['Date', 'Description', 'Totals'],\n",
    "                      'liberia': ['Date', 'Variable', 'National'],\n",
    "                      'sl': ['date', 'variable', 'National']}\n",
    "\n",
    "date_original_formats = {'guinea': ['%Y-%m-%d', '%y-%m-%d'],\n",
    "                         'liberia': ['%m/%d/%Y', '%m/%d/%y'],\n",
    "                         'sl': ['%Y-%m-%d', '%y-%m-%d']}\n",
    "dataframe_list = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read files"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for i in country_keys:\n",
    "    temp_list = []\n",
    "    all_files = glob.glob(paths[i] + '/*.csv')\n",
    "    for file in all_files:\n",
    "        temp_df = pd.read_csv(file)\n",
    "        temp_list.append(temp_df)\n",
    "    dataframe_list[i] = pd.concat(temp_list)\n",
    "    dataframe_list[i]['Country'] = countries[i]\n",
    "    dataframe_list[i][wanted_columns[i][0]] = sanitize_date(dataframe_list[i][wanted_columns[i][0]], i, date_original_formats)\n",
    "    dataframe_list[i][wanted_columns[i][0]] = pd.to_datetime(dataframe_list[i][wanted_columns[i][0]])\n",
    "    dataframe_list[i].sort_values(by=wanted_columns[i][0])\n",
    "print (dataframe_list['guinea']['Date'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in country_keys:\n",
    "    temp_list = []\n",
    "    all_files = glob.glob(paths[i] + '/*.csv')\n",
    "    for file in all_files:\n",
    "        temp_df = pd.read_csv(file)\n",
    "        temp_list.append(temp_df)\n",
    "    dataframe_list[i] = pd.concat(temp_list)\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    #dataframe_list[i][DEFAULT_NAME_COLUMN_COUNTRY] = countries[i]\n",
    "    \n",
    "    # Format dates and order dataframes by date\n",
    "    dataframe_list[i][wanted_columns[i][0]] = np.array([sanitize_date(x, date_original_formats[i])\n",
    "                                                        for x in dataframe_list[i][wanted_columns[i][0]]])\n",
    "    \n",
    "    dataframe_list[i][wanted_columns[i][0]] = pd.to_datetime(dataframe_list[i][wanted_columns[i][0]])\n",
    "    dataframe_list[i].sort_values(by=wanted_columns[i][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_guinea_files():\n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_guinea_df = dataframe_list['guinea'][wanted_columns['guinea']]\n",
    "    wanted_row_list = ['Total new cases registered so far',\n",
    "                       'New deaths registered', \n",
    "                       'New deaths registered today']\n",
    "    \n",
    "    # Select all the rows that match desired names for 'Description' column\n",
    "    parsed_guinea_df = parsed_guinea_df[parsed_guinea_df['Description'].isin(wanted_row_list)]\n",
    "    \n",
    "    # Rename every row value for 'Description' column \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[0], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[1], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[2], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_guinea_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['guinea']\n",
    "    \n",
    "    return parsed_guinea_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_liberia_files():\n",
    "    # Select certain columns from the CSV formatted files\n",
    "    liberia_df = dataframe_list['liberia'][wanted_columns['liberia']]\n",
    "\n",
    "    wanted_row_list = ['New Case/s (Suspected)',\n",
    "                       'New Case/s (Probable)',\n",
    "                       'New case/s (confirmed)',\n",
    "                       'Newly reported deaths']\n",
    "\n",
    "    # Create temporary dataframe for new case entries\n",
    "    # (with every row that matches the desired 'Variables' column value)\n",
    "    liberia_df_new_cases = liberia_df[liberia_df['Variable'].isin(wanted_row_list[0:3])]\n",
    "\n",
    "    # Sum all of the values for 'Probable', 'Variable' and 'Confirmed' new cases\n",
    "    liberia_df_new_cases = liberia_df_new_cases.groupby('Date').sum()\n",
    "\n",
    "    liberia_df_new_cases['Date'] = liberia_df_new_cases.index\n",
    "    liberia_df_new_cases['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths entries\n",
    "    liberia_df_new_deaths = liberia_df[liberia_df['Variable'] == wanted_row_list[3]]\n",
    "    liberia_df_new_deaths['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Create new dataframe with new deaths and cases (concatenating the two temporary)\n",
    "    parsed_liberia_df = pd.concat([liberia_df_new_cases, liberia_df_new_deaths])\n",
    "    \n",
    "    parsed_liberia_df.rename(columns={'Date':DEFAULT_NAME_COLUMN_DATE,\n",
    "                                      'Variable':DEFAULT_NAME_COLUMN_DESCRIPTION,\n",
    "                                      'National':DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "\n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_liberia_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['liberia']\n",
    "    \n",
    "    return parsed_liberia_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sierra_files():\n",
    "    # Select certain columns from the CSV formatted files\n",
    "    parsed_sierra_df = dataframe_list['sl'][wanted_columns['sl']]\n",
    "    \n",
    "    wanted_row_list = ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed',\n",
    "                       'death_confirmed', 'death_probable', 'death_suspected']\n",
    "    \n",
    "    # Create temporary dataframe for new case entries\n",
    "    # (with every row that matches the desired 'variables' column value)\n",
    "    sierra_df_new_cases = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[0:4])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable', 'Confirmed' and 'Noncase' new cases\n",
    "    sierra_df_new_cases = sierra_df_new_cases.groupby('date')['National'].apply(\n",
    "        lambda x: np.array([float(y) for y in x]).sum()).to_frame()\n",
    "    \n",
    "    sierra_df_new_cases['date'] = sierra_df_new_cases.index\n",
    "    sierra_df_new_cases['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths' accumulator entries\n",
    "    sierra_df_new_deaths_cum = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[4:7])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Suspected' and 'Confirmed' new death accumulators\n",
    "    sierra_df_new_deaths_cum = sierra_df_new_deaths_cum.groupby('date')['National'].apply(\n",
    "        lambda x: np.array([float(y) for y in x]).sum()).to_frame()\n",
    "    \n",
    "    sierra_df_new_deaths_cum['date'] = sierra_df_new_deaths_cum.index\n",
    "    sierra_df_new_deaths_cum['variable'] = 'New deaths accumulator'\n",
    "    \n",
    "    # \n",
    "    sierra_df_new_deaths_cum['National'] = estimate_daily_new_deaths(sierra_df_new_deaths_cum['National'])\n",
    "    sierra_df_new_deaths_cum['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    #\n",
    "    parsed_sierra_df = pd.concat([sierra_df_new_cases, sierra_df_new_deaths_cum])\n",
    "    \n",
    "    parsed_sierra_df.rename(columns={'date':DEFAULT_NAME_COLUMN_DATE,\n",
    "                                     'variable':DEFAULT_NAME_COLUMN_DESCRIPTION,\n",
    "                                     'National':DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_sierra_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['sl']\n",
    "    \n",
    "    return parsed_sierra_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_df = pd.concat([parse_guinea_files(), parse_liberia_files(), parse_sierra_files()])\n",
    "complete_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-07</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-09-20</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-02</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-11-28</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-01</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-09</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-04</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-11-30</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-08</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-08-29</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-09-05</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-09-12</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-09-30</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-10-06</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-13</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-14</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-15</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-16</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-17</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-11-19</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>New cases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Country       Date Description Totals\n",
       "137       Liberia 2014-12-03   New cases    NaN\n",
       "146       Liberia 2014-12-07  New deaths    NaN\n",
       "154       Liberia 2014-12-03  New deaths    NaN\n",
       "164       Liberia 2014-09-20  New deaths    NaN\n",
       "171       Liberia 2014-12-05  New deaths    NaN\n",
       "173       Liberia 2014-12-02  New deaths    NaN\n",
       "186       Liberia 2014-11-28  New deaths    NaN\n",
       "193       Liberia 2014-12-01  New deaths    NaN\n",
       "202       Liberia 2014-12-09  New deaths    NaN\n",
       "206       Liberia 2014-12-04  New deaths    NaN\n",
       "232       Liberia 2014-12-06  New deaths    NaN\n",
       "239       Liberia 2014-11-30  New deaths    NaN\n",
       "242       Liberia 2014-12-08  New deaths    NaN\n",
       "258  Sierra Leone 2014-08-26   New cases    NaN\n",
       "261  Sierra Leone 2014-08-29   New cases    NaN\n",
       "267  Sierra Leone 2014-09-05   New cases    NaN\n",
       "274  Sierra Leone 2014-09-12   New cases    NaN\n",
       "292  Sierra Leone 2014-09-30   New cases    NaN\n",
       "297  Sierra Leone 2014-10-06   New cases    NaN\n",
       "328  Sierra Leone 2014-11-13   New cases    NaN\n",
       "329  Sierra Leone 2014-11-14   New cases    NaN\n",
       "330  Sierra Leone 2014-11-15   New cases    NaN\n",
       "331  Sierra Leone 2014-11-16   New cases    NaN\n",
       "332  Sierra Leone 2014-11-17   New cases    NaN\n",
       "333  Sierra Leone 2014-11-18   New cases    NaN\n",
       "334  Sierra Leone 2014-11-19   New cases    NaN\n",
       "345  Sierra Leone 2014-12-06   New cases    NaN\n",
       "346  Sierra Leone 2014-12-13   New cases    NaN"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(complete_df.isnull().values.any())\n",
    "complete_df[complete_df['Totals'].isnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
