{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data/ebola' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here\n",
    "path_list = [DATA_FOLDER + '/guinea_data', DATA_FOLDER + '/liberia_data', DATA_FOLDER + '/sl_data']\n",
    "country_list = ['Guinea', 'Liberia', 'Sierra Leone']\n",
    "wanted_column_list = [['Date', 'Description', 'Totals'], ['Date', 'Variable', 'National'], ['date', 'variable', 'National']]\n",
    "dataframe_list = [pd.DataFrame(), pd.DataFrame(), pd.DataFrame()]    \n",
    "\n",
    "for i in range (0,3):\n",
    "    temp_list = []\n",
    "    all_files = glob.glob(path_list[i] + '/*.csv')\n",
    "    for file in all_files:\n",
    "        temp_df = pd.read_csv(file)\n",
    "        temp_list.append(temp_df)\n",
    "    dataframe_list[i] = pd.concat(temp_list)\n",
    "    dataframe_list[i]['Country'] = country_list[i]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nunom\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Totals</th>\n",
       "      <th>Description</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>09/23/2014</th>\n",
       "      <td>09/23/2014</td>\n",
       "      <td>88.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>09/25/2014</th>\n",
       "      <td>09/25/2014</td>\n",
       "      <td>55.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>09/26/2014</th>\n",
       "      <td>09/26/2014</td>\n",
       "      <td>71.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>09/27/2014</th>\n",
       "      <td>09/27/2014</td>\n",
       "      <td>37.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/1/2014</th>\n",
       "      <td>10/1/2014</td>\n",
       "      <td>24.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/10/2014</th>\n",
       "      <td>10/10/2014</td>\n",
       "      <td>26.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/11/14</th>\n",
       "      <td>10/11/14</td>\n",
       "      <td>31.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/12/14</th>\n",
       "      <td>10/12/14</td>\n",
       "      <td>11.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/13/14</th>\n",
       "      <td>10/13/14</td>\n",
       "      <td>36.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/16/14</th>\n",
       "      <td>10/16/14</td>\n",
       "      <td>41.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/17/14</th>\n",
       "      <td>10/17/14</td>\n",
       "      <td>167.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/18/14</th>\n",
       "      <td>10/18/14</td>\n",
       "      <td>64.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/19/14</th>\n",
       "      <td>10/19/14</td>\n",
       "      <td>19.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/20/14</th>\n",
       "      <td>10/20/14</td>\n",
       "      <td>50.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/21/2014</th>\n",
       "      <td>10/21/2014</td>\n",
       "      <td>28.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/22/2014</th>\n",
       "      <td>10/22/2014</td>\n",
       "      <td>30.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/23/2014</th>\n",
       "      <td>10/23/2014</td>\n",
       "      <td>46.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/24/14</th>\n",
       "      <td>10/24/14</td>\n",
       "      <td>35.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/25/2014</th>\n",
       "      <td>10/25/2014</td>\n",
       "      <td>31.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/28/2014</th>\n",
       "      <td>10/28/2014</td>\n",
       "      <td>55.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/29/2014</th>\n",
       "      <td>10/29/2014</td>\n",
       "      <td>69.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/3/2014</th>\n",
       "      <td>10/3/2014</td>\n",
       "      <td>58.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/30/2014</th>\n",
       "      <td>10/30/2014</td>\n",
       "      <td>44.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/31/2014</th>\n",
       "      <td>10/31/2014</td>\n",
       "      <td>32.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/4/14</th>\n",
       "      <td>10/4/14</td>\n",
       "      <td>42.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/5/14</th>\n",
       "      <td>10/5/14</td>\n",
       "      <td>28.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/7/2014</th>\n",
       "      <td>10/7/2014</td>\n",
       "      <td>56.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/8/2014</th>\n",
       "      <td>10/8/2014</td>\n",
       "      <td>56.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/9/2014</th>\n",
       "      <td>10/9/2014</td>\n",
       "      <td>60.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11/14/2014</th>\n",
       "      <td>11/14/2014</td>\n",
       "      <td>20.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/24/14</td>\n",
       "      <td>40.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/25/2014</td>\n",
       "      <td>24.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/28/2014</td>\n",
       "      <td>30.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/29/2014</td>\n",
       "      <td>20.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/30/2014</td>\n",
       "      <td>29.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/31/2014</td>\n",
       "      <td>41.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/2/2014</td>\n",
       "      <td>8.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/4/2014</td>\n",
       "      <td>24.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/8/2014</td>\n",
       "      <td>14.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/14/2014</td>\n",
       "      <td>20.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/15/2014</td>\n",
       "      <td>12.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/19/2014</td>\n",
       "      <td>21.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>30.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/21/2014</td>\n",
       "      <td>14.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/23/2014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/24/2014</td>\n",
       "      <td>15.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/26/2014</td>\n",
       "      <td>21.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/27/2014</td>\n",
       "      <td>16.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/28/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/29/2014</td>\n",
       "      <td>7.0</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/30/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/1/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/2/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/3/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/4/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/5/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/6/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/7/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/8/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/9/2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Date  Totals Description  Country\n",
       "09/23/2014  09/23/2014    88.0   New cases  Liberia\n",
       "09/25/2014  09/25/2014    55.0   New cases  Liberia\n",
       "09/26/2014  09/26/2014    71.0   New cases  Liberia\n",
       "09/27/2014  09/27/2014    37.0   New cases  Liberia\n",
       "10/1/2014    10/1/2014    24.0   New cases  Liberia\n",
       "10/10/2014  10/10/2014    26.0   New cases  Liberia\n",
       "10/11/14      10/11/14    31.0   New cases  Liberia\n",
       "10/12/14      10/12/14    11.0   New cases  Liberia\n",
       "10/13/14      10/13/14    36.0   New cases  Liberia\n",
       "10/16/14      10/16/14    41.0   New cases  Liberia\n",
       "10/17/14      10/17/14   167.0   New cases  Liberia\n",
       "10/18/14      10/18/14    64.0   New cases  Liberia\n",
       "10/19/14      10/19/14    19.0   New cases  Liberia\n",
       "10/20/14      10/20/14    50.0   New cases  Liberia\n",
       "10/21/2014  10/21/2014    28.0   New cases  Liberia\n",
       "10/22/2014  10/22/2014    30.0   New cases  Liberia\n",
       "10/23/2014  10/23/2014    46.0   New cases  Liberia\n",
       "10/24/14      10/24/14    35.0   New cases  Liberia\n",
       "10/25/2014  10/25/2014    31.0   New cases  Liberia\n",
       "10/28/2014  10/28/2014    55.0   New cases  Liberia\n",
       "10/29/2014  10/29/2014    69.0   New cases  Liberia\n",
       "10/3/2014    10/3/2014    58.0   New cases  Liberia\n",
       "10/30/2014  10/30/2014    44.0   New cases  Liberia\n",
       "10/31/2014  10/31/2014    32.0   New cases  Liberia\n",
       "10/4/14        10/4/14    42.0   New cases  Liberia\n",
       "10/5/14        10/5/14    28.0   New cases  Liberia\n",
       "10/7/2014    10/7/2014    56.0   New cases  Liberia\n",
       "10/8/2014    10/8/2014    56.0   New cases  Liberia\n",
       "10/9/2014    10/9/2014    60.0   New cases  Liberia\n",
       "11/14/2014  11/14/2014    20.0   New cases  Liberia\n",
       "...                ...     ...         ...      ...\n",
       "3             10/24/14    40.0  New deaths  Liberia\n",
       "3           10/25/2014    24.0  New deaths  Liberia\n",
       "3           10/28/2014    30.0  New deaths  Liberia\n",
       "3           10/29/2014    20.0  New deaths  Liberia\n",
       "3           10/30/2014    29.0  New deaths  Liberia\n",
       "3           10/31/2014    41.0  New deaths  Liberia\n",
       "3            11/2/2014     8.0  New deaths  Liberia\n",
       "3            11/4/2014    24.0  New deaths  Liberia\n",
       "3            11/8/2014    14.0  New deaths  Liberia\n",
       "3           11/14/2014    20.0  New deaths  Liberia\n",
       "3           11/15/2014    12.0  New deaths  Liberia\n",
       "3           11/19/2014    21.0  New deaths  Liberia\n",
       "3           11/20/2014    30.0  New deaths  Liberia\n",
       "3           11/21/2014    14.0  New deaths  Liberia\n",
       "3           11/23/2014     0.0  New deaths  Liberia\n",
       "3           11/24/2014    15.0  New deaths  Liberia\n",
       "3           11/26/2014    21.0  New deaths  Liberia\n",
       "3           11/27/2014    16.0  New deaths  Liberia\n",
       "3           11/28/2014     NaN  New deaths  Liberia\n",
       "3           11/29/2014     7.0  New deaths  Liberia\n",
       "3           11/30/2014     NaN  New deaths  Liberia\n",
       "3            12/1/2014     NaN  New deaths  Liberia\n",
       "3            12/2/2014     NaN  New deaths  Liberia\n",
       "3            12/3/2014     NaN  New deaths  Liberia\n",
       "3            12/4/2014     NaN  New deaths  Liberia\n",
       "3            12/5/2014     NaN  New deaths  Liberia\n",
       "3            12/6/2014     NaN  New deaths  Liberia\n",
       "3            12/7/2014     NaN  New deaths  Liberia\n",
       "3            12/8/2014     NaN  New deaths  Liberia\n",
       "3            12/9/2014     NaN  New deaths  Liberia\n",
       "\n",
       "[200 rows x 4 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DEFAULT_NAME_COLUMN_DATE = 'Date'\n",
    "DEFAULT_NAME_COLUMN_DESCRIPTION = 'Description'\n",
    "DEFAULT_NAME_COLUMN_TOTAL = 'Totals'\n",
    "DEFAULT_NAME_COLUMN_COUNTRY = 'Country'\n",
    "\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES = 'New cases'\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS = 'New deaths'\n",
    "\n",
    "def parse_guinea_files():\n",
    "    \n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_guinea_df = dataframe_list[0][['Date', 'Description', 'Totals']]\n",
    "    wanted_row_list = ['Total new cases registered so far', 'New deaths registered', 'New deaths registered today']\n",
    "    \n",
    "    # Select all the rows that match desired names for 'Description' column\n",
    "    parsed_guinea_df = parsed_guinea_df[parsed_guinea_df['Description'].isin(wanted_column_list)]\n",
    "    \n",
    "    # Rename every row value for 'Description' column \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[0], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[1], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[2], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_guinea_df[DEFAULT_NAME_COLUMN_COUNTRY] = 'Guinea'\n",
    "    return parsed_guinea_df\n",
    "\n",
    "def parse_liberia_files():\n",
    "    \n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_liberia_df = dataframe_list[1][['Date', 'Variable', 'National']]\n",
    "    wanted_row_list = ['New Case/s (Suspected)', 'New Case/s (Probable)', 'New case/s (confirmed)', 'Newly reported deaths']\n",
    "    \n",
    "    # Create temporary dataframe for new case entries (with every row that matches the desired 'Variables' column value)\n",
    "    parsed_liberia_df_new_cases = parsed_liberia_df[parsed_liberia_df['Variable'].isin(wanted_row_list[0:3])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable' and 'Confirmed' new deaths\n",
    "    parsed_liberia_df_new_cases = parsed_liberia_df_new_cases.groupby('Date').sum()\n",
    "    parsed_liberia_df_new_cases['Date'] = parsed_liberia_df_new_cases.index\n",
    "    parsed_liberia_df_new_cases['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths entries\n",
    "    parsed_liberia_df_new_deaths = parsed_liberia_df[parsed_liberia_df['Variable'] == wanted_row_list[3]]\n",
    "    parsed_liberia_df_new_deaths['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Create new dataframe with new deaths and cases (concatenating the two temporary)\n",
    "    parsed_liberia_df = pd.concat([parsed_liberia_df_new_cases, parsed_liberia_df_new_deaths])\n",
    "    parsed_liberia_df[DEFAULT_NAME_COLUMN_COUNTRY] = 'Liberia'\n",
    "    \n",
    "    parsed_liberia_df.rename(columns={'Date':DEFAULT_NAME_COLUMN_DATE, 'Variable':DEFAULT_NAME_COLUMN_DESCRIPTION, 'National':DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "    \n",
    "    return parsed_liberia_df\n",
    "\n",
    "parse_liberia_files()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
