{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import datetime\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data/ebola' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DEFAULT_NAME_COLUMN_DATE = 'Date'\n",
    "DEFAULT_NAME_COLUMN_DESCRIPTION = 'Description'\n",
    "DEFAULT_NAME_COLUMN_TOTAL = 'Totals'\n",
    "DEFAULT_NAME_COLUMN_COUNTRY = 'Country'\n",
    "\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES = 'New cases'\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS = 'New deaths'\n",
    "\n",
    "DEFAULT_DATA_FORMAT = '%Y-%m-%d'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auxiliary functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def sanitize_date(date_col, country_index):\n",
    "    date_list = list()\n",
    "    for i, date in date_col.iteritems():\n",
    "        date_rep = None\n",
    "        try:\n",
    "            date_rep = datetime.datetime.strptime(date, data_format_list[country_index][0])\n",
    "        except ValueError:\n",
    "            date_rep = datetime.datetime.strptime(date, data_format_list[country_index][1])\n",
    "        date_list.append(date_rep.strftime(DEFAULT_DATA_FORMAT))\n",
    "    return np.array(date_list)\n",
    "\n",
    "def sum_string_list(ls):\n",
    "    temp_sum = 0\n",
    "    for elem in ls:\n",
    "        temp_sum += float(elem)\n",
    "    return temp_sum\n",
    "    \n",
    "def estimate_daily_new_deaths(np_list):\n",
    "    last_known_n = -1\n",
    "    final_list = list()\n",
    "    for i, elem in np_list.iteritems():\n",
    "        if (math.isnan(elem)):\n",
    "            final_list.append(0)\n",
    "        else:\n",
    "            if (last_known_n == -1):\n",
    "                final_list.append(0)\n",
    "            else:\n",
    "                final_list.append(elem - last_known_n)\n",
    "            last_known_n = elem\n",
    "    return np.array(final_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_list = [DATA_FOLDER + '/guinea_data', DATA_FOLDER + '/liberia_data', DATA_FOLDER + '/sl_data']\n",
    "country_list = ['Guinea', 'Liberia', 'Sierra Leone']\n",
    "wanted_column_list = [['Date', 'Description', 'Totals'], ['Date', 'Variable', 'National'], ['date', 'variable', 'National']]\n",
    "data_format_list = [['%Y-%m-%d', '%y-%m-%d'], ['%m/%d/%Y', '%m/%d/%y'], ['%Y-%m-%d', '%y-%m-%d']]\n",
    "dataframe_list = [pd.DataFrame(), pd.DataFrame(), pd.DataFrame()]    \n",
    "\n",
    "for i in range (0,3):\n",
    "    temp_list = []\n",
    "    all_files = glob.glob(path_list[i] + '/*.csv')\n",
    "    for file in all_files:\n",
    "        temp_df = pd.read_csv(file)\n",
    "        temp_list.append(temp_df)\n",
    "    dataframe_list[i] = pd.concat(temp_list)\n",
    "    dataframe_list[i]['Country'] = country_list[i]\n",
    "    dataframe_list[i][wanted_column_list[i][0]] = sanitize_date(dataframe_list[i][wanted_column_list[i][0]], i)\n",
    "    dataframe_list[i][wanted_column_list[i][0]] = pd.to_datetime(dataframe_list[i][wanted_column_list[i][0]])\n",
    "    dataframe_list[i].sort_values(by=wanted_column_list[i][0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_guinea_files():\n",
    "    \n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_guinea_df = dataframe_list[0][['Date', 'Description', 'Totals']]\n",
    "    wanted_row_list = ['Total new cases registered so far', 'New deaths registered', 'New deaths registered today']\n",
    "    \n",
    "    # Select all the rows that match desired names for 'Description' column\n",
    "    parsed_guinea_df = parsed_guinea_df[parsed_guinea_df['Description'].isin(wanted_row_list)]\n",
    "    \n",
    "    # Rename every row value for 'Description' column \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[0], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[1], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] == wanted_row_list[2], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_guinea_df[DEFAULT_NAME_COLUMN_COUNTRY] = 'Guinea'\n",
    "    return parsed_guinea_df\n",
    "\n",
    "def parse_liberia_files():\n",
    "    \n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_liberia_df = dataframe_list[1][['Date', 'Variable', 'National']]\n",
    "    wanted_row_list = ['New Case/s (Suspected)', 'New Case/s (Probable)', 'New case/s (confirmed)', 'Newly reported deaths']\n",
    "    \n",
    "    # Create temporary dataframe for new case entries (with every row that matches the desired 'Variables' column value)\n",
    "    parsed_liberia_df_new_cases = parsed_liberia_df[parsed_liberia_df['Variable'].isin(wanted_row_list[0:3])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable' and 'Confirmed' new cases\n",
    "    parsed_liberia_df_new_cases = parsed_liberia_df_new_cases.groupby('Date').sum()\n",
    "    parsed_liberia_df_new_cases['Date'] = parsed_liberia_df_new_cases.index\n",
    "    parsed_liberia_df_new_cases['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths entries\n",
    "    parsed_liberia_df_new_deaths = parsed_liberia_df[parsed_liberia_df['Variable'] == wanted_row_list[3]]\n",
    "    parsed_liberia_df_new_deaths['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Create new dataframe with new deaths and cases (concatenating the two temporary)\n",
    "    parsed_liberia_df = pd.concat([parsed_liberia_df_new_cases, parsed_liberia_df_new_deaths])\n",
    "    parsed_liberia_df[DEFAULT_NAME_COLUMN_COUNTRY] = 'Liberia'\n",
    "    \n",
    "    parsed_liberia_df.rename(columns={'Date':DEFAULT_NAME_COLUMN_DATE, 'Variable':DEFAULT_NAME_COLUMN_DESCRIPTION, 'National':DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "    \n",
    "    return parsed_liberia_df\n",
    "\n",
    "def parse_sierra_files():\n",
    "    \n",
    "    parsed_sierra_df = dataframe_list[2][['date', 'variable', 'National']]\n",
    "    wanted_row_list = ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed', 'death_confirmed', 'death_probable', 'death_suspected']\n",
    "    \n",
    "    # Create temporary dataframe for new case entries (with every row that matches the desired 'variables' column value)\n",
    "    parsed_sierra_df_new_cases = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[0:4])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable', 'Confirmed' and 'Noncase' new cases\n",
    "    parsed_sierra_df_new_cases = parsed_sierra_df_new_cases.groupby('date')['National'].apply(lambda x: sum_string_list(x)).to_frame()\n",
    "    parsed_sierra_df_new_cases['date'] = parsed_sierra_df_new_cases.index\n",
    "    parsed_sierra_df_new_cases['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths' accumulator entries\n",
    "    parsed_sierra_df_new_deaths_cum = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[4:7])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Suspected' and 'Confirmed' new death accumulators\n",
    "    parsed_sierra_df_new_deaths_cum = parsed_sierra_df_new_deaths_cum.groupby('date')['National'].apply(lambda x: sum_string_list(x)).to_frame()\n",
    "    parsed_sierra_df_new_deaths_cum['date'] = parsed_sierra_df_new_deaths_cum.index\n",
    "    parsed_sierra_df_new_deaths_cum['variable'] = 'New deaths accumulator'\n",
    "    \n",
    "    # \n",
    "    parsed_sierra_df_new_deaths_cum['National'] = estimate_daily_new_deaths(parsed_sierra_df_new_deaths_cum['National'])\n",
    "    parsed_sierra_df_new_deaths_cum['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    #\n",
    "    parsed_sierra_df = pd.concat([parsed_sierra_df_new_cases, parsed_sierra_df_new_deaths_cum])\n",
    "    parsed_sierra_df[DEFAULT_NAME_COLUMN_COUNTRY] = \"Sierra\"\n",
    "    \n",
    "    parsed_sierra_df.rename(columns={'date':DEFAULT_NAME_COLUMN_DATE, 'variable':DEFAULT_NAME_COLUMN_DESCRIPTION, 'National':DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "    \n",
    "    return parsed_sierra_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-04</td>\n",
       "      <td>New cases</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-04</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New cases</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-27</td>\n",
       "      <td>New cases</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-27</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-30</td>\n",
       "      <td>New cases</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-30</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-31</td>\n",
       "      <td>New cases</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-31</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-02</td>\n",
       "      <td>New cases</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-02</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>New cases</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-07</td>\n",
       "      <td>New cases</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-07</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-08</td>\n",
       "      <td>New cases</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-08</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-09</td>\n",
       "      <td>New cases</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-09</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-11</td>\n",
       "      <td>New cases</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-11</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-14</td>\n",
       "      <td>New cases</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-14</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-16</td>\n",
       "      <td>New cases</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-16</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-17</td>\n",
       "      <td>New cases</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-17</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-19</td>\n",
       "      <td>New cases</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-19</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-10-27</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-10-28</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-10-30</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-10-31</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-01</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-02</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-06</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-07</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-08</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-12</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-13</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-14</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-15</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-16</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-17</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-19</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-20</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-21</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-22</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-23</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-24</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-28</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-11-29</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-12-01</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-12-04</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>Sierra</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>450 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Country       Date Description Totals\n",
       "0    Guinea 2014-08-04   New cases      9\n",
       "1    Guinea 2014-08-04  New deaths      2\n",
       "2    Guinea 2014-08-26   New cases     28\n",
       "3    Guinea 2014-08-26  New deaths      5\n",
       "4    Guinea 2014-08-27   New cases     22\n",
       "5    Guinea 2014-08-27  New deaths      2\n",
       "6    Guinea 2014-08-30   New cases     24\n",
       "7    Guinea 2014-08-30  New deaths      5\n",
       "8    Guinea 2014-08-31   New cases     46\n",
       "9    Guinea 2014-08-31  New deaths      3\n",
       "10   Guinea 2014-09-02   New cases     25\n",
       "11   Guinea 2014-09-02  New deaths      5\n",
       "12   Guinea 2014-09-04   New cases     30\n",
       "13   Guinea 2014-09-04  New deaths      5\n",
       "14   Guinea 2014-09-07   New cases     16\n",
       "15   Guinea 2014-09-07  New deaths      4\n",
       "16   Guinea 2014-09-08   New cases     16\n",
       "17   Guinea 2014-09-08  New deaths      4\n",
       "18   Guinea 2014-09-09   New cases     16\n",
       "19   Guinea 2014-09-09  New deaths      7\n",
       "20   Guinea 2014-09-11   New cases     22\n",
       "21   Guinea 2014-09-11  New deaths      2\n",
       "22   Guinea 2014-09-14   New cases     25\n",
       "23   Guinea 2014-09-14  New deaths      1\n",
       "24   Guinea 2014-09-16   New cases     10\n",
       "25   Guinea 2014-09-16  New deaths      3\n",
       "26   Guinea 2014-09-17   New cases     10\n",
       "27   Guinea 2014-09-17  New deaths      3\n",
       "28   Guinea 2014-09-19   New cases     16\n",
       "29   Guinea 2014-09-19  New deaths      5\n",
       "..      ...        ...         ...    ...\n",
       "420  Sierra 2014-10-27  New deaths     32\n",
       "421  Sierra 2014-10-28  New deaths      3\n",
       "422  Sierra 2014-10-30  New deaths      7\n",
       "423  Sierra 2014-10-31  New deaths      6\n",
       "424  Sierra 2014-11-01  New deaths      7\n",
       "425  Sierra 2014-11-02  New deaths      8\n",
       "426  Sierra 2014-11-06  New deaths     29\n",
       "427  Sierra 2014-11-07  New deaths     12\n",
       "428  Sierra 2014-11-08  New deaths      7\n",
       "429  Sierra 2014-11-10  New deaths     16\n",
       "430  Sierra 2014-11-12  New deaths     20\n",
       "431  Sierra 2014-11-13  New deaths     11\n",
       "432  Sierra 2014-11-14  New deaths      0\n",
       "433  Sierra 2014-11-15  New deaths     26\n",
       "434  Sierra 2014-11-16  New deaths     17\n",
       "435  Sierra 2014-11-17  New deaths      0\n",
       "436  Sierra 2014-11-18  New deaths     17\n",
       "437  Sierra 2014-11-19  New deaths      0\n",
       "438  Sierra 2014-11-20  New deaths     43\n",
       "439  Sierra 2014-11-21  New deaths     21\n",
       "440  Sierra 2014-11-22  New deaths     30\n",
       "441  Sierra 2014-11-23  New deaths     31\n",
       "442  Sierra 2014-11-24  New deaths     33\n",
       "443  Sierra 2014-11-28  New deaths     84\n",
       "444  Sierra 2014-11-29  New deaths     15\n",
       "445  Sierra 2014-12-01  New deaths     53\n",
       "446  Sierra 2014-12-04  New deaths     99\n",
       "447  Sierra 2014-12-05  New deaths     21\n",
       "448  Sierra 2014-12-06  New deaths     39\n",
       "449  Sierra 2014-12-13  New deaths      0\n",
       "\n",
       "[450 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_df = pd.concat([parse_guinea_files(), parse_liberia_files(), parse_sierra_files()])\n",
    "complete_df.reset_index(inplace=True, drop=True)\n",
    "complete_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
