{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import datetime\n",
    "from calendar import monthrange\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEFAULT_NAME_COLUMN_DATE = 'Date'\n",
    "DEFAULT_NAME_COLUMN_DESCRIPTION = 'Description'\n",
    "DEFAULT_NAME_COLUMN_TOTAL = 'Totals'\n",
    "DEFAULT_NAME_COLUMN_COUNTRY = 'Country'\n",
    "\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES = 'New cases'\n",
    "DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS = 'New deaths'\n",
    "\n",
    "DEFAULT_DATA_FORMAT = '%Y-%m-%d'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Auxiliary functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sanitize_date(date, original_formats):\n",
    "    '''\n",
    "    Transform a date specified in a string with a certain original\n",
    "    format into a string with the date in the default format.\n",
    "    '''\n",
    "    date_rep = None\n",
    "    try:\n",
    "        date_rep = datetime.datetime.strptime(date, original_formats[0])\n",
    "    except ValueError:\n",
    "        date_rep = datetime.datetime.strptime(date, original_formats[1])\n",
    "    return date_rep.strftime(DEFAULT_DATA_FORMAT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_daily_from_cum(np_list):\n",
    "    '''\n",
    "    Given a list of accumulated data, calculate the data for each day.\n",
    "    It assumes ordered data.\n",
    "    '''\n",
    "    last_known_n = -1\n",
    "    final_list = list()\n",
    "    for i, elem in np_list.iteritems():\n",
    "        #elem = float(elem)\n",
    "        if (math.isnan(elem)):\n",
    "            final_list.append(0)\n",
    "        else:\n",
    "            if (last_known_n == -1):\n",
    "                final_list.append(0)\n",
    "            else:\n",
    "                final_list.append(elem - last_known_n)\n",
    "            last_known_n = elem\n",
    "    return np.array(final_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define dictionaries with values for each country"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "country_keys = ['guinea', 'liberia', 'sl']\n",
    "paths = {x: DATA_FOLDER + '/ebola/'+ x + '_data' for x in country_keys}\n",
    "countries = {'guinea': 'Guinea', 'liberia': 'Liberia', 'sl': 'Sierra Leone'}\n",
    "\n",
    "wanted_columns = {'guinea': ['Date', 'Description', 'Totals'],\n",
    "                      'liberia': ['Date', 'Variable', 'National'],\n",
    "                      'sl': ['date', 'variable', 'National']}\n",
    "\n",
    "date_original_formats = {'guinea': ['%Y-%m-%d', '%y-%m-%d'],\n",
    "                         'liberia': ['%m/%d/%Y', '%m/%d/%y'],\n",
    "                         'sl': ['%Y-%m-%d', '%y-%m-%d']}\n",
    "dataframe_list = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in country_keys:\n",
    "    temp_list = []\n",
    "    all_files = glob.glob(paths[i] + '/*.csv')\n",
    "    for file in all_files:\n",
    "        temp_df = pd.read_csv(file)\n",
    "        temp_list.append(temp_df)\n",
    "    dataframe_list[i] = pd.concat(temp_list)\n",
    "    \n",
    "    # Format dates and order dataframes by date\n",
    "    dataframe_list[i][wanted_columns[i][0]] = np.array([sanitize_date(x, date_original_formats[i])\n",
    "                                                        for x in dataframe_list[i][wanted_columns[i][0]]])\n",
    "    \n",
    "    dataframe_list[i][wanted_columns[i][0]] = pd.to_datetime(dataframe_list[i][wanted_columns[i][0]])\n",
    "    dataframe_list[i].sort_values(by=wanted_columns[i][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For Guinea, we are considering the variables 'New deaths registered' and 'New deaths registered today', which were defined for all of the dates present in the data files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_guinea_files():\n",
    "    # Select certain column from the CSV formatted files\n",
    "    parsed_guinea_df = dataframe_list['guinea'][wanted_columns['guinea']]\n",
    "    wanted_row_list = ['Total new cases registered so far',\n",
    "                       'New deaths registered', \n",
    "                       'New deaths registered today']\n",
    "    \n",
    "    # Select all the rows that match desired names for 'Description' column\n",
    "    parsed_guinea_df = parsed_guinea_df[parsed_guinea_df['Description'].isin(wanted_row_list)]\n",
    "    \n",
    "    # Rename every row value for 'Description' column \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[0], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[1], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    parsed_guinea_df.loc[parsed_guinea_df['Description'] ==\n",
    "                         wanted_row_list[2], 'Description'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_guinea_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['guinea']\n",
    "    \n",
    "    return parsed_guinea_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New cases</td>\n",
       "      <td>28</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New cases</td>\n",
       "      <td>29</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>New cases</td>\n",
       "      <td>30</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Date Description Totals Country\n",
       "3 2014-08-26   New cases     28  Guinea\n",
       "9 2014-08-26  New deaths      5  Guinea\n",
       "3 2014-09-23   New cases     29  Guinea\n",
       "9 2014-09-23  New deaths      3  Guinea\n",
       "3 2014-09-04   New cases     30  Guinea"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_guinea_files().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Explain Liberia parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_liberia_files():\n",
    "    # Select certain columns from the CSV formatted files\n",
    "    liberia_df = dataframe_list['liberia'][wanted_columns['liberia']]\n",
    "\n",
    "    wanted_row_list = ['New Case/s (Suspected)',\n",
    "                       'New Case/s (Probable)',\n",
    "                       'New case/s (confirmed)',\n",
    "                       'Newly reported deaths',\n",
    "                       'Total suspected cases',\n",
    "                       'Total probable cases',\n",
    "                       'Total confirmed cases']\n",
    "\n",
    "    # Create temporary dataframe for new case entries\n",
    "    # (with every row that matches the desired 'Variables' column value)\n",
    "    liberia_df_new_cases_cum_1 = liberia_df[liberia_df['Variable'].isin(wanted_row_list[4:7])]\n",
    "    liberia_df_new_cases_cum_1 = liberia_df_new_cases_cum_1[liberia_df_new_cases_cum_1['Date'].map(\n",
    "        lambda x: x.month == 12 and x.day <= 3)]\n",
    "    liberia_df_new_cases_cum_1 = liberia_df_new_cases_cum_1.groupby('Date').sum()\n",
    "    \n",
    "    liberia_df_new_cases_cum_2 = liberia_df[liberia_df['Variable'].isin(wanted_row_list[0:3])]\n",
    "    liberia_df_new_cases_cum_2 = liberia_df_new_cases_cum_2[liberia_df_new_cases_cum_2['Date'].map(\n",
    "        lambda x: x.month == 12 and x.day > 3)]\n",
    "    liberia_df_new_cases_cum_2 = liberia_df_new_cases_cum_2.groupby('Date').sum()\n",
    "    \n",
    "    liberia_df_new_cases_cum = pd.concat([liberia_df_new_cases_cum_1, liberia_df_new_cases_cum_2])\n",
    "\n",
    "    \n",
    "    liberia_df_new_cases_cum['National'] = estimate_daily_from_cum(liberia_df_new_cases_cum['National'])\n",
    "    \n",
    "    liberia_df_new_cases = liberia_df[liberia_df['Variable'].isin(wanted_row_list[0:3])]\n",
    "    liberia_df_new_cases = liberia_df_new_cases[liberia_df_new_cases['Date'].map(\n",
    "        lambda x: x.month != 12)]\n",
    "    \n",
    "    liberia_df_new_cases = pd.concat([liberia_df_new_cases, liberia_df_new_cases_cum])\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable' and 'Confirmed' new cases\n",
    "    liberia_df_new_cases = liberia_df_new_cases.groupby('Date').sum()\n",
    "\n",
    "    liberia_df_new_cases['Date'] = liberia_df_new_cases.index\n",
    "    liberia_df_new_cases['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths entries\n",
    "    liberia_df_new_deaths = liberia_df[liberia_df['Variable'] == wanted_row_list[3]]\n",
    "    liberia_df_new_deaths['Variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    # Create new dataframe with new deaths and cases (concatenating the two temporary)\n",
    "    parsed_liberia_df = pd.concat([liberia_df_new_cases, liberia_df_new_deaths])\n",
    "    \n",
    "    parsed_liberia_df.rename(columns={'Date': DEFAULT_NAME_COLUMN_DATE,\n",
    "                                      'Variable': DEFAULT_NAME_COLUMN_DESCRIPTION,\n",
    "                                      'National': DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "\n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_liberia_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['liberia']\n",
    "    \n",
    "    return parsed_liberia_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Totals</th>\n",
       "      <th>Description</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-06-16 00:00:00</th>\n",
       "      <td>2014-06-16</td>\n",
       "      <td>4.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-17 00:00:00</th>\n",
       "      <td>2014-06-17</td>\n",
       "      <td>2.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-22 00:00:00</th>\n",
       "      <td>2014-06-22</td>\n",
       "      <td>10.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-24 00:00:00</th>\n",
       "      <td>2014-06-24</td>\n",
       "      <td>6.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-25 00:00:00</th>\n",
       "      <td>2014-06-25</td>\n",
       "      <td>7.0</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Date  Totals Description  Country\n",
       "2014-06-16 00:00:00 2014-06-16     4.0   New cases  Liberia\n",
       "2014-06-17 00:00:00 2014-06-17     2.0   New cases  Liberia\n",
       "2014-06-22 00:00:00 2014-06-22    10.0   New cases  Liberia\n",
       "2014-06-24 00:00:00 2014-06-24     6.0   New cases  Liberia\n",
       "2014-06-25 00:00:00 2014-06-25     7.0   New cases  Liberia"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_liberia_files().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Explain sierrra parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sierra_files():\n",
    "    # Select certain columns from the CSV formatted files\n",
    "    parsed_sierra_df = dataframe_list['sl'][wanted_columns['sl']]\n",
    "    \n",
    "    wanted_row_list = ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed',\n",
    "                       'death_confirmed', 'death_probable', 'death_suspected']\n",
    "    \n",
    "    # Create temporary dataframe for new case entries\n",
    "    # (with every row that matches the desired 'variables' column value)\n",
    "    sierra_df_new_cases = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[0:4])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Variable', 'Confirmed' and 'Noncase' new cases\n",
    "    sierra_df_new_cases = sierra_df_new_cases.groupby('date')['National'].apply(\n",
    "        lambda x: np.array([float(y) for y in x]).sum()).to_frame()\n",
    "    \n",
    "    sierra_df_new_cases['date'] = sierra_df_new_cases.index\n",
    "    sierra_df_new_cases['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES\n",
    "    \n",
    "    # Create temporary dataframe for new deaths' accumulator entries\n",
    "    sierra_df_new_deaths_cum = parsed_sierra_df[parsed_sierra_df['variable'].isin(wanted_row_list[4:7])]\n",
    "    \n",
    "    # Sum all of the values for 'Probable', 'Suspected' and 'Confirmed' new death accumulators\n",
    "    sierra_df_new_deaths_cum = sierra_df_new_deaths_cum.groupby('date')['National'].apply(\n",
    "        lambda x: np.array([float(y) for y in x]).sum()).to_frame()\n",
    "    \n",
    "    sierra_df_new_deaths_cum['date'] = sierra_df_new_deaths_cum.index\n",
    "    sierra_df_new_deaths_cum['variable'] = 'New deaths accumulator'\n",
    "    \n",
    "    # \n",
    "    sierra_df_new_deaths_cum['National'] = estimate_daily_from_cum(sierra_df_new_deaths_cum['National'])\n",
    "    sierra_df_new_deaths_cum['variable'] = DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS\n",
    "    \n",
    "    #\n",
    "    parsed_sierra_df = pd.concat([sierra_df_new_cases, sierra_df_new_deaths_cum])\n",
    "    \n",
    "    parsed_sierra_df.rename(columns={'date': DEFAULT_NAME_COLUMN_DATE,\n",
    "                                     'variable': DEFAULT_NAME_COLUMN_DESCRIPTION,\n",
    "                                     'National': DEFAULT_NAME_COLUMN_TOTAL}, inplace=True)\n",
    "    \n",
    "    # Add 'Country' column to dataframe\n",
    "    parsed_sierra_df[DEFAULT_NAME_COLUMN_COUNTRY] = countries['sl']\n",
    "    \n",
    "    return parsed_sierra_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Totals</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-08-12</th>\n",
       "      <td>26.0</td>\n",
       "      <td>2014-08-12</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-13</th>\n",
       "      <td>32.0</td>\n",
       "      <td>2014-08-13</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-14</th>\n",
       "      <td>41.0</td>\n",
       "      <td>2014-08-14</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-15</th>\n",
       "      <td>32.0</td>\n",
       "      <td>2014-08-15</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-16</th>\n",
       "      <td>30.0</td>\n",
       "      <td>2014-08-16</td>\n",
       "      <td>New cases</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Totals       Date Description       Country\n",
       "date                                                   \n",
       "2014-08-12    26.0 2014-08-12   New cases  Sierra Leone\n",
       "2014-08-13    32.0 2014-08-13   New cases  Sierra Leone\n",
       "2014-08-14    41.0 2014-08-14   New cases  Sierra Leone\n",
       "2014-08-15    32.0 2014-08-15   New cases  Sierra Leone\n",
       "2014-08-16    30.0 2014-08-16   New cases  Sierra Leone"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_sierra_files().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We join the three together:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_df = pd.concat([parse_guinea_files(), parse_liberia_files(), parse_sierra_files()])\n",
    "complete_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New cases</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New cases</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>New cases</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Country       Date Description Totals\n",
       "0  Guinea 2014-08-26   New cases     28\n",
       "1  Guinea 2014-08-26  New deaths      5\n",
       "2  Guinea 2014-09-23   New cases     29\n",
       "3  Guinea 2014-09-23  New deaths      3\n",
       "4  Guinea 2014-09-04   New cases     30"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have some NaN values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-07</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-09-20</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Liberia</td>\n",
       "      <td>2014-12-02</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Country       Date Description Totals\n",
       "137  Liberia 2014-12-07  New deaths    NaN\n",
       "145  Liberia 2014-12-03  New deaths    NaN\n",
       "155  Liberia 2014-09-20  New deaths    NaN\n",
       "162  Liberia 2014-12-05  New deaths    NaN\n",
       "164  Liberia 2014-12-02  New deaths    NaN"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_df[complete_df['Totals'].isnull()].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th>Description</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New cases</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New cases</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>New deaths</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>New cases</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Country       Date Description Totals\n",
       "0  Guinea 2014-08-26   New cases     28\n",
       "1  Guinea 2014-08-26  New deaths      5\n",
       "2  Guinea 2014-09-23   New cases     29\n",
       "3  Guinea 2014-09-23  New deaths      3\n",
       "4  Guinea 2014-09-04   New cases     30"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assume NaN values are 0\n",
    "complete_df = complete_df.fillna(0)\n",
    "complete_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have seen that the difference in the number of deaths between 2014-09-30 and 2014-10-01 is negative. We decided to ignore this negative number of deaths, since it is ...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Maybe remove the row with -18 deaths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate for each country, the daily average per month of new cases and deaths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = complete_df.groupby(by=[complete_df.Country, complete_df.Description, [x.month for x in complete_df.Date]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "calculated_average_df = pd.DataFrame()\n",
    "\n",
    "for ((country, description, month), values) in grouped:\n",
    "    days_in_month = monthrange(2014, month)[1]\n",
    "    entry = {'Country':country, 'Description':description, 'Month':month, 'Average':values['Totals'].apply(lambda x: float(x) / days_in_month).sum()}\n",
    "    calculated_average_df = calculated_average_df.append([entry])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "death_average_df = calculated_average_df[calculated_average_df['Description'] == DEFAULT_NAME_ROW_DESCRIPTION_NEW_DEATHS]\n",
    "cases_average_df = calculated_average_df[calculated_average_df['Description'] == DEFAULT_NAME_ROW_DESCRIPTION_NEW_CASES]\n",
    "\n",
    "calculated_average_df = pd.merge(death_average_df, cases_average_df, on=['Country', 'Month'])\n",
    "calculated_average_df.drop(['Description_x', 'Description_y'], axis=1, inplace=True)\n",
    "calculated_average_df = calculated_average_df.rename(columns={'Average_x': 'Death monthly average',\n",
    "                                                              'Average_y': 'Cases monthly average'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Death monthly average</th>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th>Cases monthly average</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.548387</td>\n",
       "      <td>Guinea</td>\n",
       "      <td>8</td>\n",
       "      <td>4.161290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.900000</td>\n",
       "      <td>Guinea</td>\n",
       "      <td>9</td>\n",
       "      <td>10.466667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>Guinea</td>\n",
       "      <td>10</td>\n",
       "      <td>1.096774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.466667</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>6</td>\n",
       "      <td>1.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.516129</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>7</td>\n",
       "      <td>3.032258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6.741935</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>8</td>\n",
       "      <td>10.806452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>28.833333</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>9</td>\n",
       "      <td>51.066667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>22.612903</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>10</td>\n",
       "      <td>36.741935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6.733333</td>\n",
       "      <td>Liberia</td>\n",
       "      <td>11</td>\n",
       "      <td>13.233333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4.258065</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>8</td>\n",
       "      <td>22.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5.433333</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>9</td>\n",
       "      <td>64.866667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>27.000000</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>10</td>\n",
       "      <td>110.903226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14.233333</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>11</td>\n",
       "      <td>98.533333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>6.838710</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>12</td>\n",
       "      <td>21.903226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Death monthly average       Country  Month  Cases monthly average\n",
       "0                0.548387        Guinea      8               4.161290\n",
       "1                1.900000        Guinea      9              10.466667\n",
       "2                0.483871        Guinea     10               1.096774\n",
       "3                0.466667       Liberia      6               1.333333\n",
       "4                1.516129       Liberia      7               3.032258\n",
       "5                6.741935       Liberia      8              10.806452\n",
       "6               28.833333       Liberia      9              51.066667\n",
       "7               22.612903       Liberia     10              36.741935\n",
       "8                6.733333       Liberia     11              13.233333\n",
       "9                4.258065  Sierra Leone      8              22.483871\n",
       "10               5.433333  Sierra Leone      9              64.866667\n",
       "11              27.000000  Sierra Leone     10             110.903226\n",
       "12              14.233333  Sierra Leone     11              98.533333\n",
       "13               6.838710  Sierra Leone     12              21.903226"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calculated_average_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fb678ea3d30>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAFZCAYAAACWmOQIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl8VOW9x/HPL4AGRQHZREGDXpBddqGogIhacUOl2OJe\ntSgFwdaKVivtbW/dbou7xQUoekUb3FfQQkErlh1ZVBRQUIGUSlAEWfK7f5yTMAkJS2YmmXPyfb9e\neWXmzMz3PCThN8885znPMXdHRETiK6uyGyAiIumlQi8iEnMq9CIiMadCLyIScyr0IiIxp0IvIhJz\nKvQiIjGnQi8iEnMq9CIiMVe9shsAUL9+fc/JyansZoiIRMrcuXP/7e4N9va8jCj0OTk5zJkzp7Kb\nISISKWb22b48T0M3IiIxp0IvIhJzKvQiIjGXEWP0pdm+fTtr1qxh69atld0UibHs7GyaNGlCjRo1\nKrspImmTsYV+zZo1HHLIIeTk5GBmld0ciSF3Z8OGDaxZs4ZmzZpVdnNE0iZjh262bt1KvXr1VOQl\nbcyMevXq6VOjxF7GFnpARV7STn9jUhVkdKEXEZHkZewYfUk5o15Nad6qO/rv9TnVqlWjXbt2bN++\nnerVq3PppZcycuRIsrL2//1x48aN/N///R/XXXcdANOnT+eee+7hlVde2e+sVBg/fjynnXYaRxxx\nBLDrpLX69evv9rw5c+bwwAMPVEYzRfbP6NplbM+v2HZkGPXo96BmzZosWLCAJUuWMHXqVF5//XV+\n+9vflitr48aNPPTQQyluYfmNHz+eL7/8srKbsU927NhR2U0QiTQV+n3UsGFDxo4dywMPPIC7s3Pn\nTm688Ua6du1K+/bt+ctf/gLAt99+S9++fenUqRPt2rXjxRdfBGDUqFF8+umndOjQgRtvvLHouRde\neCEtW7Zk8ODBuPtu++3duzcjR46kS5cutGrVitmzZ3P++efTvHlzbr311qLn/elPf6Jt27a0bduW\nMWPGALBq1SpatWrF1VdfTZs2bTjttNPYsmULubm5zJkzh8GDB9OhQwe2bNkCwP3331/U7g8//LBY\nO7755huaNWvG9u3bAdi0aVOx+4VefvllTjjhBDp27Mipp57KunXrKCgoICcnh40bNxY9r3nz5qxb\nt468vDwuuOACunbtSteuXXn33XcBGD16NJdccgk9e/bkkksuYdWqVZx00kl06tSJTp068c9//hOA\ngoICrrvuOlq2bEm/fv0488wzyc3NBWDu3Ln06tWLzp07c/rpp/PVV1+V51cvEnkq9PvhmGOOYefO\nnaxfv57HH3+c2rVrM3v2bGbPns2jjz7KypUryc7O5vnnn2fevHlMmzaNX/ziF7g7d9xxB8ceeywL\nFizg7rvvBmD+/PmMGTOGpUuXsmLFiqIiV9IBBxzAnDlzGDJkCOeeey4PPvggixcvZvz48WzYsIG5\nc+cybtw43n//fWbNmsWjjz7K/PnzAVi+fDlDhw5lyZIl1KlTh8mTJ3PhhRfSpUsXnnrqKRYsWEDN\nmjUBqF+/PvPmzePaa6/lnnvuKdaGQw45hN69e/Pqq8EQ2qRJkzj//PN3m39+4oknMmvWLObPn89F\nF13EXXfdRVZWFueeey7PP/88AO+//z5HH300jRo14vrrr2fkyJHMnj2byZMnc9VVVxVlLV26lLfe\neounn36ahg0bMnXqVObNm8czzzzD8OHDAXjuuedYtWoVS5cuZeLEibz33ntAcB7GsGHDyM3NZe7c\nuVx55ZX8+te/Tur3LxJVkRmjzzRTpkxh0aJFRb3H/Px8li9fTpMmTbjllluYMWMGWVlZfPHFF6xb\nt67UjG7dutGkSRMAOnTowKpVqzjxxBN3e94555wDQLt27WjTpg2NGzcGgjee1atX88477zBgwAAO\nPvhgAM4//3xmzpzJOeecQ7NmzejQoQMAnTt3ZtWqVWX+m84///yi5z333HO7PX7VVVdx1113cd55\n5zFu3DgeffTR3Z6zZs0aBg0axFdffcW2bduK5qcPGjSI3/3ud1xxxRVMmjSJQYMGAfDWW2+xdOnS\notdv2rSJb7/9tujfXfgmtH37dn7+85+zYMECqlWrxscffwzAO++8w8CBA8nKyuLwww+nT58+AHz0\n0UcsXryYfv36AbBz586in5tIVaNCvx9WrFhBtWrVaNiwIe7O/fffz+mnn17sOePHjycvL4+5c+dS\no0YNcnJyypynfeCBBxbdrlatWplj0YXPy8rKKvaarKysvY5fl9xH4TDNnp5bVlt69uzJqlWrmD59\nOjt37qRt27a7PWfYsGHccMMNnHPOOUyfPp3Ro0cD0KNHDz755BPy8vJ44YUXioadCgoKmDVrFtnZ\n2btlFb5xAfz5z3+mUaNGLFy4kIKCglKfn8jdadOmTVEPX6Qq09DNPsrLy2PIkCH8/Oc/x8w4/fTT\nefjhh4vGqD/++GM2b95Mfn4+DRs2pEaNGkybNo3PPgtWET3kkEP45ptv0tK2k046iRdeeIHvvvuO\nzZs38/zzz3PSSSft8TXlbc+ll17KT37yE6644opSH8/Pz+fII48EYMKECUXbzYwBAwZwww030KpV\nK+rVqwfAaaedxv3331/0vAULFpSZ27hxY7Kyspg4cSI7d+4EgjefyZMnU1BQwLp165g+fToAxx13\nHHl5ecWGcpYsWbLf/16ROIhMj35fpkOm2pYtW+jQoUPR9MpLLrmEG264AQiGMVatWkWnTp1wdxo0\naMALL7zA4MGDOfvss2nXrh1dunShZcuWANSrV4+ePXvStm1bfvjDH9K/f+r+PZ06deLyyy+nW7du\nRW3r2LHjHodpLr/8coYMGULNmjX3q9c7ePBgbr31Vn784x+X+vjo0aMZOHAgdevW5ZRTTmHlypVF\njw0aNIiuXbsyfvz4om333XcfQ4cOpX379uzYsYOTTz6ZRx55ZLfc6667jgsuuIC//vWvnHHGGUW9\n/QsuuIC3336b1q1b07RpUzp16kTt2rU54IADyM3NZfjw4eTn57Njxw5GjBhBmzZt9vnfKhIXVtpM\nj4rWpUsXL3nhkWXLltGqVatKapGUJTc3lxdffJGJEydWdlOKfPvtt9SqVYsNGzbQrVs33n33XQ4/\n/PB9fr3+1mKkis2jN7O57t5lb8+LTI9eKt+wYcN4/fXXee211yq7KcWcddZZbNy4kW3btnHbbbft\nV5EXqQpU6GWfJY6lZ5LCcXkRKZ0OxoqIxJwKvYhIzKnQi4jEnAq9iEjMRedgbFnTpsqdt/fpVmvX\nrmXEiBHMnj2bOnXq0KhRI8aMGUOLFi1S25YKtmDBAr788kvOPPNMIJj7XqtWLX75y1/u9txatWoV\nLUkgItGkHn0Z3J0BAwbQu3dvPv30U+bOncsf//jHMtetiZIFCxZk3BTJsrg7BQUFld0MkUjba6E3\nsyfMbL2ZLU7YdpiZTTWz5eH3ugmP3Wxmn5jZR2Z2eumpmW/atGnUqFGDIUOGFG07/vjjOemkk8pc\ninjz5s3079+f448/nrZt2/LMM88AZS+Xe99999G6dWvat2/PRRddtFsbxo8fz3nnnUe/fv3Iycnh\ngQce4E9/+hMdO3ake/fu/Oc//wGCwt29e3fat2/PgAED+Prrr4FgieObbrqJbt260aJFC2bOnMm2\nbdv4zW9+wzPPPEOHDh2K2rh06VJ69+7NMcccw3333bdbWy699FJeeOGFovuDBw8u+ncX2tMSzQ8+\n+GDR80aPHl20Oubdd99dtNTz7bffDgTLKx933HFceumltG3bltWrV3PttdfSpUsX2rRpU/Q8gNde\ne42WLVvSuXNnhg8fzllnnVX0u7jyyivp1q0bHTt23K2tIlWKu+/xCzgZ6AQsTth2FzAqvD0KuDO8\n3RpYCBwINAM+BartbR+dO3f2kpYuXVp8w+2HpvZrL+69914fMWJEqY9t377d8/Pz3d09Ly/Pjz32\nWC8oKPDc3Fy/6qqrip63ceNG37Ztm/fo0cPXr1/v7u6TJk3yK664wt3dGzdu7Fu3bnV396+//nq3\n/YwbN86PPfZY37Rpk69fv94PPfRQf/jhh93dfcSIEf7nP//Z3d3btWvn06dPd3f32267za+//np3\nd+/Vq5ffcMMN7u7+6quvet++fYtyhw4duutHe/vt3qNHD9+6davn5eX5YYcd5tu2bXN394MPPtjd\n3adPn+7nnntu0b8rJyfHt2/fvk8/l3nz5vnJJ59c9LxWrVr5559/7m+++aZfffXVXlBQ4Dt37vT+\n/fv7P/7xD1+5cqWbmb/33ntFr9mwYYO7u+/YscN79erlCxcu9C1btniTJk18xYoV7u5+0UUXef/+\n/d3d/eabb/aJEycW/WybN2/u3377bam/z93+1iS6yvn/PaqAOb6X+urue+/Ru/sM4D8lNp8LFK5Y\nNQE4L2H7JHf/3t1XAp8A3cr1DpTB3J1bbrmF9u3bc+qppxYtRdyuXTumTp3KTTfdxMyZM6ldu3ax\n5XI7dOjA73//e9asWQNA+/btGTx4ME8++STVq5d+uKRPnz4ccsghNGjQgNq1a3P22WcDwZLFq1at\nIj8/n40bN9KrVy8ALrvsMmbMmFH0+sSlh/e09k3//v058MADqV+/Pg0bNtxtiKpXr14sX76cvLw8\nnn76aS644ILd2lzWz6Vjx46sX7+eL7/8koULF1K3bl2aNm3KlClTmDJlCh07dqRTp058+OGHLF++\nHICjjz6a7t27F2U/++yzdOrUiY4dO7JkyRKWLl3Khx9+yDHHHFO0FHLi+jtTpkzhjjvuoEOHDvTu\n3ZutW7fy+eefl/1LFYmx8h6MbeTuhZfrWQs0Cm8fCcxKeN6acNtuzOwa4BqAo446qpzNSJ82bdoU\nrTVf0lNPPVXqUsQtWrRg3rx5vPbaa9x666307duXAQMGlLlc7quvvsqMGTN4+eWX+cMf/sAHH3yw\nW/EsuSxx4pLF+3KJvb0tPVzafsp67qWXXsqTTz7JpEmTGDdu3G6Pl/VzARg4cCC5ubmsXbu2aC16\nd+fmm2/mZz/7WbGcVatWFVuieOXKldxzzz3Mnj2bunXrcvnll5e59HMhd2fy5Mkcd9xxe3yeSFWQ\n9MHY8OPDfq+M5u5j3b2Lu3dp0KBBss1IuVNOOYXvv/+esWPHFm1btGgRM2fOLHMp4i+//JKDDjqI\niy++mBtvvJF58+aVuVxuQUEBq1evpk+fPtx5553k5+eXa3ZL7dq1qVu3LjNnzgRg4sSJRb37spR3\nieLLL7+86DKFrVu33u3xsn4uEKxcOWnSJHJzcxk4cCAAp59+Ok888UTRv/uLL75g/fr1u+Vu2rSJ\ngw8+mNq1a7Nu3Tpef/11IFiKeMWKFUWfVAqPNxRm33///UWXZyy84pZIVVTeHv06M2vs7l+ZWWOg\n8H/nF0DThOc1Cbclr4JXnzMznn/+eUaMGMGdd95JdnY2OTk5jBkzpsyliD/44ANuvPFGsrKyqFGj\nBg8//HCZy+W2aNGCiy++mPz8fNyd4cOHU6dOnXK1dcKECQwZMoTvvvuOY445ptTedqI+ffoUDWvc\nfPPN+7yfRo0a0apVK84777xSHy/r5wLBJ6RvvvmGI488suhKT6eddhrLli2jR48eQDCV88knn6Ra\ntWrFco8//ng6duxIy5Ytadq0KT179gSCi7c/9NBDRcsWd+3ateg1t912GyNGjKB9+/YUFBTQrFkz\nXnnllX3+t4rEyT4tU2xmOcAr7t42vH83sMHd7zCzUcBh7v4rM2sD/B/BuPwRwNtAc3ffuad8LVMc\nDd999x3t2rVj3rx51K6d4vMayqlwiWJ3Z+jQoTRv3pyRI0fuV4b+1mIk3csUZ9gyyPu6TPG+TK98\nGngPOM7M1pjZT4E7gH5mthw4NbyPuy8BngWWAm8AQ/dW5CUa3nrrLVq1asWwYcMypsgDPProo3To\n0IE2bdqQn5+/23i/iOzD0I27l34pIehbxvP/APwhmUZJ5jn11FOLjblnipEjR+53D16kqsnoM2P3\nZVhJJBn6G5OqIGMLfXZ2Nhs2bNB/REkbd2fDhg1kZ2dXdlNE0ipjFzVr0qQJa9asIS8vr7KbIjGW\nnZ1NkyZNKrsZImmVsYW+Ro0aRWc8iohI+WXs0I2IiKSGCr2ISMyp0IuIxJwKvYhIzKnQi4jEnAq9\niEjMqdCLiMScCr2ISMyp0IuIxJwKvYhIzKnQi4jEnAq9iEjMqdCLiMScCr2ISMyp0IuIxJwKvYhI\nzKnQi4jEnAq9iEjMqdCLiMScCr2ISMyp0IuIxJwKvYhIzKnQi4jEnAq9iEjMqdCLiMRcUoXezEaa\n2RIzW2xmT5tZtpkdZmZTzWx5+L1uqhorIiL7r9yF3syOBIYDXdy9LVANuAgYBbzt7s2Bt8P7IiJS\nSZIduqkO1DSz6sBBwJfAucCE8PEJwHlJ7kNERJJQ7kLv7l8A9wCfA18B+e4+BWjk7l+FT1sLNCrt\n9WZ2jZnNMbM5eXl55W2GiIjsRTJDN3UJeu/NgCOAg83s4sTnuLsDXtrr3X2su3dx9y4NGjQobzNE\nRGQvkhm6ORVY6e557r4deA74AbDOzBoDhN/XJ99MEREpr2QK/edAdzM7yMwM6AssA14CLgufcxnw\nYnJNFBGRZFQv7wvd/X0zywXmATuA+cBYoBbwrJn9FPgM+FEqGioiIuVT7kIP4O63A7eX2Pw9Qe9e\nREQygM6MFRGJORV6EZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmFOhFxGJORV6EZGYU6EXEYk5FXoR\nkZhToRcRiTkVehGRmFOhFxGJORV6EZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmFOhFxGJORV6EZGY\nU6EXEYk5FXoRkZirXtkNEJEqZHTtMrbnV2w7qhj16EVEYk6FXkQk5lToRURiToVeRCTmVOhFRGIu\nqUJvZnXMLNfMPjSzZWbWw8wOM7OpZrY8/F43VY0VEZH9l2yP/l7gDXdvCRwPLANGAW+7e3Pg7fC+\niIhUknIXejOrDZwMPA7g7tvcfSNwLjAhfNoE4LxkGykiIuWXTI++GZAHjDOz+Wb2mJkdDDRy96/C\n56wFGpX2YjO7xszmmNmcvLy8JJohIiJ7kkyhrw50Ah52947AZkoM07i7A17ai919rLt3cfcuDRo0\nSKIZIiKyJ8kU+jXAGnd/P7yfS1D415lZY4Dw+/rkmigiIskod6F397XAajM7LtzUF1gKvARcFm67\nDHgxqRaKiEhSkl3UbBjwlJkdAKwAriB483jWzH4KfAb8KMl9iIhIEpIq9O6+AOhSykN9k8kVEZHU\n0ZmxIiIxp0IvIhJzKvQiIjGnQi8iEnMq9CIiMadCLyIScyr0IiIxp0IvIhJzKvQiIjGnQi8iEnMq\n9CIiMZfsomYi0TO6dhnb8yu2HSIVRD16EZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmFOhFxGJORV6\nEZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmFOhFxGJORV6EZGYU6EXEYk5FXoRkZhToRcRibmkC72Z\nVTOz+Wb2Snj/MDObambLw+91k2+miIiUVyp69NcDyxLujwLedvfmwNvhfRERqSRJFXozawL0Bx5L\n2HwuMCG8PQE4L5l9iIhIcpLt0Y8BfgUUJGxr5O5fhbfXAo1Ke6GZXWNmc8xsTl5eXpLNEBGRspS7\n0JvZWcB6d59b1nPc3QEv47Gx7t7F3bs0aNCgvM0QEZG9qJ7Ea3sC55jZmUA2cKiZPQmsM7PG7v6V\nmTUG1qeioSIiUj7l7tG7+83u3sTdc4CLgL+7+8XAS8Bl4dMuA15MupUiIlJu6ZhHfwfQz8yWA6eG\n90VEpJIkM3RTxN2nA9PD2xuAvqnIFRGR5OnMWBGRmEtJj15EEoyuXcb2/Ipth0hIPXoRkZhToRcR\niTkVehGRmFOhFxGJORV6EZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmFOhFxGJORV6EZGYU6EXEYk5\nFXoRkZjT6pUSWzmjXi11+6rsCm6ISCVTj15EJOZU6EVEYk6FXkQk5jRGLyKRo+Mv+0c9ehGRmFOh\nFxGJORV6EZGYU6EXEYk5FXoRkZhToRcRiTkVehGRmNM8epFyiu1c7tG1S9mWX/HtkJRRoRcRyRSl\nvclC0m+05R66MbOmZjbNzJaa2RIzuz7cfpiZTTWz5eH3ukm1UEREkpLMGP0O4Bfu3hroDgw1s9bA\nKOBtd28OvB3eFxGRSlLuQu/uX7n7vPD2N8Ay4EjgXGBC+LQJwHnJNlJERMovJbNuzCwH6Ai8DzRy\n96/Ch9YCjcp4zTVmNsfM5uTl5aWiGSIiUoqkC72Z1QImAyPcfVPiY+7ugJf2Oncf6+5d3L1LgwYN\nkm2GiIiUIalCb2Y1CIr8U+7+XLh5nZk1Dh9vDKxProkiIpKMZGbdGPA4sMzd/5Tw0EvAZeHty4AX\ny988ERFJVjLz6HsClwAfmNmCcNstwB3As2b2U+Az4EfJNVFERJJR7kLv7u8AVsbDfcubKyIiqaW1\nbkREYk6FXkQk5lToRURiToVeRCTmVOhFRGJOyxSLiJQQt2sNqNCLRE2a1iyX+NLQjYhIzKnQi4jE\nnAq9iEjMqdCLiMScDsZK5tHBRpGUUqEvjQqNiMSIhm5ERGJOhV5EJOZU6EVEYk6FXkQk5lToRURi\nTrNuREQqWEUvmqZCL5UmbisEppp+PpIqGroREYk59ehFJOX0aSSzqEcvIhJz0ezRa4kCkaSp1111\nqEcvIhJzKvQiIjEXzaEbqVwaOhOJFPXoRURiLm09ejM7A7gXqAY85u53pGtf5aWDUXumn49IPKSl\n0JtZNeBBoB+wBphtZi+5+9J07C9y0j30oaEVEUmQrh59N+ATd18BYGaTgHOB/Sr0Ue9Rprv9Uf/5\niEjFMHdPfajZhcAZ7n5VeP8S4AR3/3nCc64BrgnvHgd8tB+7qA/8O0XNVb7ylV9x+VFueybmH+3u\nDfb2pEqbdePuY4Gx5Xmtmc1x9y4pbpLyla/8NOdHue1Rzk/XrJsvgKYJ95uE20REpIKlq9DPBpqb\nWTMzOwC4CHgpTfsSEZE9SMvQjbvvMLOfA28STK98wt2XpHAX5RryUb7ylV/p+VFue2Tz03IwVkRE\nMofOjBURiTkVehGRmFOhFxGJuUitXmlmdYEjgC3AKncviFJ+OplZQ6Anu9q/GJiTin9DOrPjkB91\n+vnHX8YfjDWz2sBQ4MfAAUAekA00AmYBD7n7tEzNT9hPF+Akiv+xT3X3r5PM7QOMAg4D5gPrCdrf\nAjgWyAX+1903ZVJ2HPLDfTQhmD5c8nf7KvB6ssUsnfn6+cc/v2g/ESj0U4G/Ai+7+8YSj3UGLgE+\ncPfHMzT/CmAYsBKYS/E/9p4Ev9Tb3P3zcubfDdxf2uvNrDpwFlDN3SdnUnZM8scBRwKvAHMo/rvt\nA3QGRrn7jAzN188/xvnF9pXphT7qzGwowXkEW8p4vANQz93frtiWSbLMrK27L97D4wcAR7n7J5mY\nH3VR//lX5O83UoU+HENvTvCuB0Aq3u0qKj/dzKw/0Ibi7f9dpmfHIT/KzKwR8D/Ake5+hpm1BnqU\n91NsReeH+6hJUBT3Z3HEKpMfmVk3ZnYVMIPgbNvfht9HRyg/28yGmtlDZvZE4VcK8x8BBhEMExkw\nEDg607Njkt/czHLNbKmZrSj8iko+MJ7g771xeP9jYERU8s3sHGAB8EZ4v4OZpWzJlajnA+DukfgC\nPiDojS0I77cEnotQ/t+A/wY+BS4DpgD3pjB/UYnvtYCZmZ4dk/x3gL7AIoI3kNHA7yKUPzv8Pj9h\n24II5c8FapfI/0D5u74i06MHtrr7VgAzO9DdPyRYxz4q+f/l7rcBm919AtAfOCGF+YXHAL4zsyOA\n7ezqQWVydhzya3pwjMXc/TN3H03w+41K/mYzqwc4gJl1B1J5ObJ0529395J5qRyTjnp+pObRrzGz\nOsALwFQz+xr4LEL528PvG82sLbAWaJjC/FfC9t8NzCP4Q3ksAtlxyP/ezLKA5eFifl8QfGqISv4N\nBKvLHmtm7wINgAsjlL/EzH4CVDOz5sBw4J/K3yVSB2MLmVkvgo86b7j7tijkh8cAJgPtgXEE/1F/\n4+6PpCK/xL4OBLJL6SVkdHZU882sK7AMqEMwPFcbuMvdZ0UhP9xHdYJPsAZ85O7b9/KSjMk3s4OA\nXwOnhflvAv9d+Am9qudDxAq9mZ0INHf3cWbWAKjl7iujkp8OZnaKu//dzM4v7XF3fy4Ts+OQHydm\n9gMgh4RP+e7+16jky55FZujGzG4HuhD0CsYBNYAnCU46ikJ+4RSzI9z9hymcYtYL+DtwdimPOZBM\nMUtnduTzzWyMu48ws5cpZUzV3c/J5PyE/UwkOFN1AbCzMJ7gRMIo5LcAfsnubySnKD/cR1R69Ga2\nAOgIzHP3juG2Re7ePiL5rxO8gfza3Y8PP8rOd/d2KcjOAi5092eTzarI7Kjnm1lnd58bDvXtxt3/\nkcn5CftZBrT2NBWDCshfCDxCMHul8I0Ed5+r/EBkevTANnd3Mys8cn9wxPLru/uzZnYzFF2Fa+fe\nXrQv3L3AzH4FpLyYpTM76vlhEa4GXOPug6OWn2AxcDjwVUTzd7j7w2nKjkN+pAr9s2b2F6COmV0N\nXAk8GqH8dE8xe8vMfgk8A2wu3Oju/8nw7Ejnu/tOMzvazA5Ix8SAdOeH6gNLzexfwPcJ+07J0FAF\n5L9sZtcBz5fIT9XfT9TzozN0A2Bm/Ug4Mu3uU6OSb2adgPuBtgQ9nAYEQwqLUpRf2kFjd/djMjk7\nJvl/BVoRTCFMfCP5U0Ty0z00lO78qP/9pDUfIlbooy7dU9ikcoQH8nfj7r+NQn64j0ZA1/Duv9x9\nfaqyKyJf9iwyhT6cIncnwUlGFn65ux8ahfxwH2mbYhbOxb2BYGGka8ITL45z91cyOTsO+Yn7cffv\nUplZEflm9iOCk8mmE/zdnwTc6O65EcmvAVwLnBxumg78JVUdqajnQ7QK/SfA2e6+LKL5pU4xc/fh\nKcp/huA16VqsAAAUN0lEQVSo/aXu3jYsbv909w6ZnB2T/B7A4wTnXRxlZscDP3P36yKSvxDoV9jL\nDs8hecvdj49I/mME06EnhJsuAXa6+1XKD3kKF85J5xfwbsTzlxG+saYpf074PXFhpIWZnh2T/PeB\npiXyF0co/4MS97NKbsvw/N1+lyn+/UY6390jNetmTtgze4HiR6ZTdXZjuvPTPcVsmwVrWhfO6jmW\nhH9HBmfHIR93X21miZtSMnW2gvLfMLM3gafD+4OA1yOUv9PMjnX3TwHM7BhS+/OJen6kCv2hwHcE\ns2IKpeLsyYrKT/cUs9sJ1rNuamZPEZzRe3kEsuOQvzo8/uLheOv1BJ/gIpHv7jeGx6hODDeNdffn\no5IP3AhMs2CNfiNYyvkK5e8SmTH6qEv3FLNwH/WA7gR/LLPc/d9RyI56vpnVB+4FTiUYlngTuN7d\nN0QhP9zH0QTrPL0VHsOo5u7fRCj/QHYtK/6Ru6f0E1vU8zO+R29mv3L3u8zsfkpf7yOpg5npzk/I\nSVlB34NeBL0mJzi4k8peUzqzI50fvmmk7czVdOeHJwheAxxGMGHgSIJT8vtGJL9wVtXR7n61BVfk\nSsesrUjmQzQuJVj4EXUOwcyJkl8ZnW9m74TfvzGzTQlf35jZpmTzE/bzEDCE4EpZi4GfmdmDmZ4d\nk/xjzOxlM8szs/Vm9mI4zhqJfGAowXDWJgB3X05qr5WQ7vxxwDagR3j/C+D3yk+QyiO7+qq8L+BD\nEmb1ELyJL8v07JjkzyKYElc9/LoYeD9C+e+H3+eH36sTXnYxIvlRn7WV1nz3CM26MbNplD60kqql\nQtOdf1Rp293981TkA58AR7HrqlhNw22Znh2H/IPcfWLC/SfN7MYI5f/DzG4BalqwDMh1wMsRyo/6\nrK20zwqLTKEnWK+5UDZwAbAjQvmvlshvBnwEtEkm1HatVX4IsCyc1eME16P9V6ZmxyT/sPDm62Y2\nCpgU5g8CXsv0/ASjgJ8SDG39DHjN3VO5oF+686M+ayvd+dGedWNm/3L3blHMt2CRs+s8ybPfyprN\nU8iTOAiczuyY5K8kKLxWysPuSS5Kle78vez7GXcfFJX8krOqgAPc/Uvlh/lRKfQJvRsIxlg7A/e5\n+3FlvCSj8svY5weegguPiKSamX3u7qUONyo/evlRGrqZy67ezQ5gJcHHwUjkm9kNCXezgE5A0u/Y\nZvaOu59oZt9Q/BhD0ouypTM7Jvm65m3mKu1TUJXNj0yhd/dmUc4nGCcutINgzH5ysqHufmL4/ZC9\nPTeTsuOQT/SveduprIcIzjVISrrz9yLdQxWRys/4oRszOxRo5MHcW8xsIFAzfPhNd1+XyfmVKZ0f\nL6P20bUS8i9w96TfyNOZH840K5O798nw/FJPciR4I7ksBZ/YIp1fLDAChX4swZKy48P7nxAsiFST\n4FqLQzI8/0TgGA/XnTezXIIzBAF+7+5/TyZ/L/te7e5No5Ydk/xIv1FFgZldtqfH3X3Cnh6Pe36i\nKAzddCWYklXoG3cfBrvOOs3w/N8CwxLuH0cwdepg4BaCj+bpks538Uh9dK2E/EiN4UZRKgthHPMT\nRaHQV/fiHzsuSbhdJwL5h7r70oT7y919LoCZ/THZ8BIHeYs9BNTK1Ow45O9F1N+oJEaiUOgLzOxw\nd18L4O6LAczsSKAgAvnF3izcPXEGRaMU5O/pQOO9GZwd+Xwz+4Cyx1iT/t2mO1+qjiiM0V9MsP72\nL4D54eZOwD0E89wnlvXaDMl/GXjE3V8tsf0s4Fp3759MvlQeC5beLZO7f7anxys7v8S+6gLNCc7a\nLsyfEZV82bOML/QAZnYGwXh2G4IezhLgDndPyVVq0plvZv9FMJXyn8C8cHNn4AfAWe7+cbL7EEmG\nmV1F0NlpQnBN4+7Aeylc5ynd+dkE57y0ofgbyZXKD0RhmWLc/Q13P9nd67l7fXfvlaoin+58d/8E\naA/MBHLCrxlAexV5yRDXE0xK+Cyc8tgR2Bih/IkEl+k8HfgHwRtKyi5qEoP8aPToRSR9zGy2u3c1\nswXACe7+vZktcfekFtyrwPz57t7RzBa5e3sLLrc40927Kz8QhYOxUg5mdi6w1t3fj1J2HPIjaI2Z\n1QFeAKaa2dfsWtI5Cvnbw+8bzawtsJbUXtgk6vkq9DF2AtDOzKq7+w8jlB35fDObQHCh+QcLZ3Fl\ncr67Dwhvjg7PZq1NsGxuSqQ7HxgbHuy9FXiJYOrsbcrfJbJDN+r1SaYys64EFzrp5u43ZXK+mVUD\nlrh7y5Q0ruLzs4AL3f1Z5e9hPxEu9P8DtCM44SkdvbJ056ekV2bBlavWu/tWMzOCs247AUuBR929\n3BdPMbNzgCnuvrW8Gfuwj5OBde7+kZn1JLhu5rKS01GTyK8FnEFwVamdwMcE/6ZUnCMRC2b2IjDM\nU3e1s4rOn+PuXdKRHYd8iHChj7pU9crMbHGY8Z2Z3QkcSzAWegokN0XLzLYAmwnW/nmaYJG3neXN\nKyV/DNCNYAjxTaBvuK9eBNfPTOpyeWb2I4Irhy0C+hBMcc0ieAO/2N0XJZlfnWBa3ADgiHDzF8CL\nwOPuvr2s1ybLzMa6+zUpyppBMBPmXwS/bwDc/ZyI5N8B/Bt4pkT+f5Qf7iMKhV69vrKZ2VJ3bx3e\nngt0LWy3mS109+OTyJ5P8IZxIXAR0BZ4Hnjak7w6U5i/JMysSVAgjwzfsGoQFPq2SeYvArqHmfWB\np9z9dDNrT3AS2w+SzH+aYJrgBGBNuLkJcBlwmCd5BSUrfjGcYg8RXDy6STL5Cfsp9UpcqfgdV1D+\nytLjU3MFrqjnQwQOxib2+swssdc30sx6p6nXdwJwl5lFode32sILVACrCN6sPrPg0mTJcnf/GngU\neNTMDgd+BNxhZk08+dUf3d3dzArfUAt7HQWk5hwPA7aEtzcTzmRw90UWLE+drM7u3qLEtjXALDNL\nxTkSeQSzUxIXMCu8OE5KZmWEY+ijPcklgysxP4vg09m7yt8Dd8/oL4KzVA04CPgaOCjcXgNYnIL8\nRQmZ9QmGJyA4yemfKch/GniY4GzAJuFX93DbMynIbwpMIzgJ6+XwZzSNYDmHvklmz9/DY0enoO13\nEpxINhu4O2z/r4EpBD3uVOS/GWbOBG4Jtx9GcIAw2fxZwEAgK2FbFsHFu99PQf5y4KgyHludbH5C\n1ttA7VTlVUJ+mX+nyg++Mr5Hj3p9e+Tuq4E+ZtYKaAGMD/Nne/JDTyP3sN+k50G7+01m1iO46bPM\n7FiCTz6PAbkpyj8TaA38zt2nhg9tJDhgnayLCN5MHgrnhkOwiN208LFkjQHqAqUdxLwrBfmFvgU+\nMLOpFB8jHh6R/LfN7ALgOQ8rZ4pFPT/zx+jDA4w/IFgDYjrQkqAn1QtY4clfGOROoANBj/gM4HV3\n/59wfHSmJ3n2npnNAv4XmOy7xs6zCHqCN7j7CUnm297+OPblORWdHYf8Ejn1ANx9Q7JZFc3KuACG\np2i99ArI/4bg+g47CTptKbkmcFzyIQKFHqCMXt/nQG4Keq0k9PoWFvb6wmJcw92/TzI7h6DXdwrB\nsArs6vWNcvfSDsTsT/50gmvPvugJ09fM7ADgRIIDg9M8vIJWpmTHIX8v++6X8Aki5VKdb2Y1CYaJ\nPkpVZkXmy55lfKFXr2+vmdnAlcBgoBnBsEQ2UI1grPshd59fdkLlZO8hvybBkFy68lPW/r3sOzKX\nEjSzswmW5T7A3ZuZWQeCoa5UTX9Md74R/o7d/b/NrCnQ2N3/pfxwHxEo9NNRr29f82oQHFDe4u6p\nXB0wrdlRzTezl8p6CDjF3Q/O5PyE/cwl+MQ53d07htsWe5LTWysw/2GCY3anuHsrC5YTmOLuXZUf\niMLB2DMIemVPm1lpvb4xSfbKSstP7PUlm78njxOcNJUSHkzV/CpVeRWVHeH8k4CLCQ42JjKCKcGZ\nnl9ou7vnBx3LIqk8hyTd+Se4eycLzvvA3b8OO2rKD2V8offg9PuHCGY2pLxXlu78vfTKUjHXXSrP\nLOA7L+XEHzNLxVh0uvMLLTGznwDVzKw5MJzgfJKo5G+3YL6+A5hZA1L8RhXx/Mwfuom6cNpdWb2y\nZ9xd1/6USmVmBxGca3Aawd/lm8B/e4rWOKqA/MEE5y50IjhL+ULgVnf/m/LDfajQp5eZvQ7c5e7T\nSnlshrufXAnNkhSI00SBqDOzlgRnzRvwtrsvU35Cvv5GRMon6hMFzGyMu4+w4AL2uxWCZGfFVED+\noe6+ycpYE8iTXBQs6vnF9qVCn17qlcVXJU0/TWV+Z3efa2ladKwC8l9x97MsWBSscA2ghPjkFgWL\nen6xfam+pFdlTt+UihPF6aFl7KMt8IW7r49avpQtFWvFyJ6dQXBq89Nm9qWZLTWzFQQLVv2YYPrm\n+MpsoCTP3be7+1fpKsLpyDezR8ysTXi7NrAQ+Csw38x+HIH8o8Pcwvt9zOxeMxuZiumJUc9PpEKf\nZu6+1d0fcveewNEEB1w6ufvR7n51Gufoi+zNSe6+JLx9BfCxu7cDOgO/ikD+swRrxGDB2bZ/I1ga\npQPBlOmqnl8k4+fRx0m6TwoS2U/bEm73Iyg0uPvaEic3ZWp+TXf/Mrx9MfCEu/+vBetULVD+LurR\ni1RdG83sLDPrCPQE3gCw4GI5NSOQn/hucQrBuvd46q4MF/X8IurRi1RdPwPuAw4HRrj72nB7XyAV\nl+lMd/7fzexZgk/JdYG/A5hZY4p/mqiq+UU060ZEIsmC8Z9BQGPgWXf/ItzeEWjo7m9W5fxi+1Kh\nFxGJN43Ri4jEnAq9iEjMqdCLSDFmdq6ZJXUt48rMl91p1o2IlHQC0M7Mqrv7D6OWb2YTgO+AB919\nsfJ1MFZEYsbMuhJcua2bu9+kfBV6kSrNzE4G1rn7R2bWE+gBLHP3VMxzx8xqEaz31JRgzaePCa6H\nmvKTgqRsKvQiVZSZjSG49mx1gqs+9QVeB3oB8939xiTzfwT8ElgE9CG4fGAW0A642N0XJZlfHfgp\nMAA4Itz8BfAi8Hi45EhamNlYd78myYxqwFVAE+ANd3834bFb3f33STZz175U6EWqJjNbQrBscE2C\nAnmku38XLic8393bJpm/COgeZtYHnnL3082sPfCIu/8gyfynCdbonwCsCTc3IVj6+zB3H5RkfqkX\nBCFYumChuzdJMv8x4CDgX8AlwD/c/YbwsXnu3imZ/EQ6GCtSdbm7u5kVDqMU9voKSM2MPAO2hLc3\nAw3DnS4ys0NTkN/Z3VuU2LYGmGVmH6cgPw/4jBIXBAnvN0xBfjd3bw9gZg8AD5nZcwTLl6dk1bdC\nKvQiVderZjaT4KpVjwHPmtksgqGbGSnIfw14w8xmEIzT/w2KesqpKGT/MbOBwOTCMf9w5ceBwNcp\nyF8B9E28YFAhM1udgvyiNefdfQdwjZn9hmDNm1opyC+ioRuRKszMehD07GeZ2bEE492fA7mpOGBq\nZmcCrQmGOqaG27KAGu7+fZLZOcCdBCs/Fhb2OsA0YJS7r0wyfyjwjrsvLOWxYe5+f5L5TwJPuvsb\nJbZfBTzs7jWSyS+WqUIvUjWZpfd6xunOL5FTD8DdNySbFUc6M1ak6ppmZsPM7KjEjWZ2gJmdEp64\nc1kG5xdx9w2JRd7M+qUityxRy1ePXqSKMrNs4EpgMNCMYAZLTYIO4BTgoWQudVlGfjZQLRX5e9n3\n5+5+1N6fWTXyVehFhHBKZX1gSzoucJ6OfDN7qayHgFPc/eCqnF8sUIVeRKLIzL4muNbqtyUfAp5x\n90ZVOT+RpleKSFTNAr5z93+UfMDMPlJ+Qp569CIi8aZZNyISSeE1V5N+TlzzE6nQi0hURX16aIVN\nP9XQjYhEUrqnb0Y9v9i+VOhFJOqiOD20QvNV6EVE4k1j9CIiMadCLyIScyr0knHMzMMlXAvvVzez\nPDN7pZx5dczsuoT7vfc1y8x+aWYfmtkCM5ttZpeWpw372jaRdFChl0y0GWhrZjXD+/0ILnVXXnWA\n/S6mZjYk3Hc3d+9AcE3VlF75Z09tC6+JKpI0FXrJVK8B/cPbPwaeLnzAzA4zsxfMbJGZzQqvQYqZ\njTazJ8xsupmtMLPh4UvuAI4Ne+V3h9tqmVlu2Ft/qowTU24BrnX3TQDuvsndJ4T76mtm883sg3Cf\nB4bbV1lwfVTMrIuZTd+ftoWfNmaGC14tNbPfmdmIhH/7H8zs+uR+tFLluLu+9JVRXwSLPLUHcgnm\nFS8AegOvhI/fD9we3j4FWBDeHg38EziQYKraBqAGkAMsTsjvDeQTXEg6C3gPOLFEGw4Fvi6jfdnA\naqBFeP+vwIjw9iqgfni7CzC9HG3bDDQL7+cA88LbWcCnQL3K/h3pK1pf6tFLRnL3RQRF7scEvftE\nJwITw+f9Hahnuy42/aq7f+/u/wbWA2WtAPgvd1/jweXyFoT72lfHASvdvfAC1BOAk/fhdfvTtpUA\n7r4K2GBmHYHTgPmuqyjJftIYoGSyl4B7CHq59fbxNYnXId1J2X/je3yeu28ys2/N7Bh3X7GP+wbY\nwa4h0exytm1zifuPAZcDhwNP7EdbRACN0UtmewL4rbt/UGL7TILTxjGz3sC/PRxHL8M3wCH7skMz\n+6OZDQjv/hF4sPDTgpnVCmfdfATkmNl/hc+7BChcanYV0Dm8fcE+7HJf2vY8cAbQFXhzX/4dIolU\n6CVjhUMr95Xy0Gigs5ktIjiYuceFn8KhjnfNbHHCwdiytAPWhrcfBqYBs81sMcEbTIG7bwWuAP5m\nZh8ABcAj4Wt+C9xrZnMIeu17+zfutW3uvi1sx7PuvtdMkZK0BIJIAjN7091Pr+x2JDKzLGAeMNDd\nl1d2eyR61KMXSZCBRb418Anwtoq8lJd69CIiMacevYhIzKnQi4jEnAq9iEjMqdCLiMScCr2ISMz9\nP1JYnMohP3dUAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fb678ea32e8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "calculated_average_df.plot.bar(x=['Month','Country'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>BARCODE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanopyri Methanopyr...</td>\n",
       "      <td>10</td>\n",
       "      <td>MID4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0   1 BARCODE\n",
       "0  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...   2    MID4\n",
       "1  Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...   1    MID4\n",
       "2  Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...   1    MID4\n",
       "3  Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...   1    MID4\n",
       "4  Archaea \"Euryarchaeota\" Methanopyri Methanopyr...  10    MID4"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_files = glob.glob(DATA_FOLDER + '/microbiome' + '/MID*.xls')\n",
    "raw_df = pd.DataFrame()\n",
    "\n",
    "for file in all_files:\n",
    "    file_name = file[file.rfind('/')+1:file.rfind('.')]\n",
    "    temp_df = pd.read_excel(file, header=None)\n",
    "    temp_df['BARCODE'] = file_name\n",
    "    raw_df = raw_df.append(temp_df)\n",
    "\n",
    "raw_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MID3</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MID4</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MID5</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MID7</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BARCODE               GROUP  SAMPLE\n",
       "0    MID1  EXTRACTION CONTROL     NaN\n",
       "1    MID2               NEC 1  tissue\n",
       "2    MID3           Control 1  tissue\n",
       "3    MID4               NEC 2  tissue\n",
       "4    MID5           Control 2  tissue\n",
       "5    MID6               NEC 1   stool\n",
       "6    MID7           Control 1   stool\n",
       "7    MID8               NEC 2   stool\n",
       "8    MID9           Control 2   stool"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_excel(DATA_FOLDER+'/microbiome'+'/metadata.xls')\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_df = pd.merge(raw_df, metadata_df[['BARCODE', 'GROUP']], on=['BARCODE'])\n",
    "raw_df.rename(columns={0: 'Description', 1: 'Sample'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "stool_metadata_df = metadata_df[metadata_df['SAMPLE'] == 'stool']\n",
    "tissue_metadata_df = metadata_df[metadata_df['SAMPLE'] == 'tissue']\n",
    "na_metadata_df = metadata_df[metadata_df['SAMPLE'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MID3</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MID4</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MID5</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BARCODE      GROUP  SAMPLE\n",
       "1    MID2      NEC 1  tissue\n",
       "2    MID3  Control 1  tissue\n",
       "3    MID4      NEC 2  tissue\n",
       "4    MID5  Control 2  tissue"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "stool_df = raw_df[raw_df['BARCODE'].isin(list(stool_metadata_df['BARCODE']))].drop('BARCODE', axis=1)\n",
    "stool_df.rename(columns={'Sample': 'Stool'}, inplace=True)\n",
    "tissue_df = raw_df[raw_df['BARCODE'].isin(list(tissue_metadata_df['BARCODE']))].drop('BARCODE', axis=1)\n",
    "tissue_df.rename(columns={'Sample': 'Tissue'}, inplace=True)\n",
    "na_df = raw_df[raw_df['BARCODE'].isin(list(na_metadata_df['BARCODE']))].drop('BARCODE', axis=1)\n",
    "na_df.rename(columns={'Sample': 'Other'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Stool</th>\n",
       "      <th>GROUP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7</td>\n",
       "      <td>NEC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>NEC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>NEC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>NEC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>5</td>\n",
       "      <td>NEC 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Description  Stool  GROUP\n",
       "134  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      7  NEC 1\n",
       "135  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1  NEC 1\n",
       "136  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1  NEC 1\n",
       "137  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      2  NEC 1\n",
       "138  Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...      5  NEC 1"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stool_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Stool</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Thermopro...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Thermopro...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanobacteria Methan...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanopyri Methanopyr...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Thermoplasmata Thermop...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Bifid...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Corio...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria \"Aquificae\" Aquificae Aquificales Aqu...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria \"Aquificae\" Aquificae Aquificales Aqu...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria \"Aquificae\" Aquificae Aquificales Des...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1491</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Deltaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1492</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Deltaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1493</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Deltaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1494</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Deltaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1495</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Deltaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1496</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Epsilonproteobacteri...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1497</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1499</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1501</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1502</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1504</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1505</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1507</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1509</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1510</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1515</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>Bacteria \"Synergistetes\" Synergistia Synergist...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>Bacteria TM7    TM7_genera_incertae_sedis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1521 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Description  Stool  GROUP  Tissue\n",
       "0     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...    7.0  NEC 1    23.0\n",
       "1     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...    1.0  NEC 1     1.0\n",
       "2     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...    1.0  NEC 1     NaN\n",
       "3     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...    2.0  NEC 1     2.0\n",
       "4     Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...    5.0  NEC 1    10.0\n",
       "5     Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...    4.0  NEC 1    11.0\n",
       "6     Archaea \"Crenarchaeota\" Thermoprotei Thermopro...    4.0  NEC 1     9.0\n",
       "7     Archaea \"Crenarchaeota\" Thermoprotei Thermopro...    1.0  NEC 1     5.0\n",
       "8     Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...   12.0  NEC 1     9.0\n",
       "9     Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...    2.0  NEC 1     1.0\n",
       "10    Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...    1.0  NEC 1    12.0\n",
       "11    Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...    1.0  NEC 1     NaN\n",
       "12    Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...    1.0  NEC 1     2.0\n",
       "13    Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...    4.0  NEC 1     4.0\n",
       "14    Archaea \"Euryarchaeota\" Halobacteria Halobacte...    4.0  NEC 1     4.0\n",
       "15    Archaea \"Euryarchaeota\" Halobacteria Halobacte...    1.0  NEC 1     NaN\n",
       "16    Archaea \"Euryarchaeota\" Halobacteria Halobacte...    3.0  NEC 1     5.0\n",
       "17    Archaea \"Euryarchaeota\" Halobacteria Halobacte...    1.0  NEC 1     NaN\n",
       "18    Archaea \"Euryarchaeota\" Halobacteria Halobacte...    3.0  NEC 1     2.0\n",
       "19    Archaea \"Euryarchaeota\" Methanobacteria Methan...    1.0  NEC 1     1.0\n",
       "20    Archaea \"Euryarchaeota\" Methanopyri Methanopyr...   23.0  NEC 1    47.0\n",
       "21    Archaea \"Euryarchaeota\" Thermoplasmata Thermop...   11.0  NEC 1    10.0\n",
       "22    Bacteria \"Actinobacteria\" Actinobacteria Actin...    1.0  NEC 1     NaN\n",
       "23    Bacteria \"Actinobacteria\" Actinobacteria Actin...    1.0  NEC 1     NaN\n",
       "24    Bacteria \"Actinobacteria\" Actinobacteria Actin...    1.0  NEC 1    31.0\n",
       "25    Bacteria \"Actinobacteria\" Actinobacteria Bifid...    1.0  NEC 1     NaN\n",
       "26    Bacteria \"Actinobacteria\" Actinobacteria Corio...    5.0  NEC 1     2.0\n",
       "27    Bacteria \"Aquificae\" Aquificae Aquificales Aqu...   35.0  NEC 1    54.0\n",
       "28    Bacteria \"Aquificae\" Aquificae Aquificales Aqu...    1.0  NEC 1     NaN\n",
       "29    Bacteria \"Aquificae\" Aquificae Aquificales Des...    5.0  NEC 1     7.0\n",
       "...                                                 ...    ...    ...     ...\n",
       "1491  Bacteria \"Proteobacteria\" Deltaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1492  Bacteria \"Proteobacteria\" Deltaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1493  Bacteria \"Proteobacteria\" Deltaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1494  Bacteria \"Proteobacteria\" Deltaproteobacteria ...    NaN  NEC 1     2.0\n",
       "1495  Bacteria \"Proteobacteria\" Deltaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1496  Bacteria \"Proteobacteria\" Epsilonproteobacteri...    NaN  NEC 1     3.0\n",
       "1497  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1498  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     2.0\n",
       "1499  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1500  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1501  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1502  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1503  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1504  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     2.0\n",
       "1505  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1506  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1507  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1508  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1509  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1510  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1511  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     9.0\n",
       "1512  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1    10.0\n",
       "1513  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     9.0\n",
       "1514  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     1.0\n",
       "1515  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     2.0\n",
       "1516  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    NaN  NEC 1     3.0\n",
       "1517  Bacteria \"Synergistetes\" Synergistia Synergist...    NaN  NEC 1     2.0\n",
       "1518  Bacteria Cyanobacteria Cyanobacteria  Chloropl...    NaN  NEC 1     1.0\n",
       "1519  Bacteria Cyanobacteria Cyanobacteria  Chloropl...    NaN  NEC 1     2.0\n",
       "1520          Bacteria TM7    TM7_genera_incertae_sedis    NaN  NEC 1     2.0\n",
       "\n",
       "[1521 rows x 4 columns]"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(stool_df, tissue_df, how='outer', on=['Description', 'GROUP'])\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = pd.merge(merged_df, na_df, how='outer', on=['Description', 'GROUP'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.fillna('unknown', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
